{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total Number of projects in GDC portal: 39\n",
      "###################################################\n",
      "TARGET-NBL\tNeuroblastoma\n",
      "TCGA-BRCA\tBreast Invasive Carcinoma\n",
      "TARGET-AML\tAcute Myeloid Leukemia\n",
      "TARGET-WT\tHigh-Risk Wilms Tumor\n",
      "TCGA-GBM\tGlioblastoma Multiforme\n",
      "TCGA-OV\tOvarian Serous Cystadenocarcinoma\n",
      "TCGA-LUAD\tLung Adenocarcinoma\n",
      "TCGA-UCEC\tUterine Corpus Endometrial Carcinoma\n",
      "TCGA-KIRC\tKidney Renal Clear Cell Carcinoma\n",
      "TCGA-HNSC\tHead and Neck Squamous Cell Carcinoma\n",
      "TCGA-LGG\tBrain Lower Grade Glioma\n",
      "TCGA-THCA\tThyroid Carcinoma\n",
      "TCGA-LUSC\tLung Squamous Cell Carcinoma\n",
      "TCGA-PRAD\tProstate Adenocarcinoma\n",
      "TCGA-STAD\tStomach Adenocarcinoma\n",
      "TCGA-SKCM\tSkin Cutaneous Melanoma\n",
      "TCGA-COAD\tColon Adenocarcinoma\n",
      "TCGA-BLCA\tBladder Urothelial Carcinoma\n",
      "TARGET-OS\tOsteosarcoma\n",
      "TCGA-LIHC\tLiver Hepatocellular Carcinoma\n",
      "TCGA-CESC\tCervical Squamous Cell Carcinoma and Endocervical Adenocarcinoma\n",
      "TCGA-KIRP\tKidney Renal Papillary Cell Carcinoma\n",
      "TCGA-SARC\tSarcoma\n",
      "TCGA-LAML\tAcute Myeloid Leukemia\n",
      "TCGA-PAAD\tPancreatic Adenocarcinoma\n",
      "TCGA-ESCA\tEsophageal Carcinoma\n",
      "TCGA-PCPG\tPheochromocytoma and Paraganglioma\n",
      "TCGA-READ\tRectum Adenocarcinoma\n",
      "TCGA-TGCT\tTesticular Germ Cell Tumors\n",
      "TCGA-THYM\tThymoma\n",
      "TCGA-KICH\tKidney Chromophobe\n",
      "TCGA-ACC\tAdrenocortical Carcinoma\n",
      "TCGA-MESO\tMesothelioma\n",
      "TCGA-UVM\tUveal Melanoma\n",
      "TARGET-RT\tRhabdoid Tumor\n",
      "TCGA-DLBC\tLymphoid Neoplasm Diffuse Large B-cell Lymphoma\n",
      "TCGA-UCS\tUterine Carcinosarcoma\n",
      "TCGA-CHOL\tCholangiocarcinoma\n",
      "TARGET-CCSK\tClear Cell Sarcoma of the Kidney\n",
      "###################################################\n",
      "\n",
      "What type is the data?\n",
      "(1.mRNA 2.miRNA 3.isoform 4.methylation27 5.methylation450 6.copy number 7.masked copy number 8.clinical 9.biospecimen 10.snp):\n",
      "3\n"
     ]
    }
   ],
   "source": [
    "######################################################\n",
    "######## Step0. Set up Environment Variables #########\n",
    "######################################################\n",
    "\n",
    "### GDC DATA PORTAL: https://gdc.cancer.gov/\n",
    "\n",
    "import re, os, scipy, json\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "from os import system\n",
    "from collections import OrderedDict, defaultdict\n",
    "from operator import itemgetter\n",
    "from pprint import pprint\n",
    "from xml.etree.ElementTree import parse\n",
    "from datetime import datetime\n",
    "import string\n",
    "\n",
    "\n",
    "easyTCGA = '/home/ruidong/Documents/Research/Cancer_Genomics/easyTCGA/'       ###### !!!!!!!!\n",
    "databases = '/home/ruidong/Documents/Research/Cancer_Genomics/Data/Databases/'\n",
    "\n",
    "### data type <not in use>\n",
    "RNAseq,miRNA,isoform,methylation27,methylation450,CNS,maskedCNS,MaskedSM,clinical,biospecimen = \\\n",
    "['Gene_Expression_Quantification/','miRNA_Expression_Quantification/','Isoform_Expression_Quantification/',\n",
    " 'Methylation_Beta_Value27/','Methylation_Beta_Value450/','Copy_Number_Segment/','Masked_Copy_Number_Segment/',\n",
    " 'Masked_Somatic_Mutation/','Clinical_Supplement/','Biospecimen_Supplement/']\n",
    "\n",
    "\n",
    "### working directory\n",
    "miRNAmRNA = 'miRNA_mRNA_network/'\n",
    "survival = 'Survival/'\n",
    "\n",
    "downloads = 'Downloads/'\n",
    "clean = 'Clean/'\n",
    "analysis = 'Analysis/'\n",
    "\n",
    "os.chdir(easyTCGA)\n",
    "\n",
    "### projects summary, add more information\n",
    "projects = OrderedDict()\n",
    "with open('projects.2017-02-03T22-29-28.039066.txt', 'rt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        if line.startswith('ID'):\n",
    "            continue\n",
    "        projects[lst[0]] = lst[1]\n",
    "        \n",
    "\n",
    "print ('Total Number of projects in GDC portal: ' + str(len(projects)))\n",
    "print ('###################################################')\n",
    "for ke, val in projects.items():\n",
    "    print (ke+'\\t'+val)\n",
    "    \n",
    "print ('###################################################\\n')\n",
    "\n",
    "genoDataType = int(input('What type is the data?\\n\\\n",
    "(1.mRNA 2.miRNA 3.isoform 4.methylation27 5.methylation450 6.copy number 7.masked copy number 8.clinical 9.biospecimen 10.snp):\\n'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "###################################################\n",
    "############# Data organization ###################\n",
    "###################################################\n",
    "\n",
    "############### targets #########################\n",
    "\n",
    "### miRNA, isoform, cnv, RNAseq\n",
    "\n",
    "\n",
    "uuid2sample = OrderedDict()\n",
    "\n",
    "metaProjects = OrderedDict()\n",
    "uuid = OrderedDict()\n",
    "samples = OrderedDict()\n",
    "traits = OrderedDict()\n",
    "\n",
    "dataType = maskedCNS\n",
    "samType = 'masked_cnv'\n",
    "\n",
    "\n",
    "with open(easyTCGA+samType+'.metadata.cart.2017-02-06.all.basic_information.txt', 'rt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        \n",
    "        lst = [trt.replace(' ','') for trt in lst]\n",
    "        if line.startswith('file_name'):\n",
    "            traitName = lst[4:16]\n",
    "            #fh.write('\\t'+'\\t'.join(lst[4:16])+'\\n')\n",
    "            continue\n",
    "            \n",
    "        elif lst[-1] not in metaProjects.keys():\n",
    "            metaProjects[lst[-1]] = []\n",
    "            \n",
    "        metaProjects[lst[-1]].append(lst[3]+'\\t'+'\\t'.join(lst[4:16]))\n",
    "            \n",
    "\n",
    "for ke, val in metaProjects.items():\n",
    "    fh = open(ke+'/'+dataType+analysis+samType+'.targets.'+ke+'.txt', 'wt')\n",
    "    fh.write('\\t'+'\\t'.join(traitName)+'\\n')\n",
    "    fh.write('\\n'.join(sorted(val)))\n",
    "    \n",
    "    fh.close()\n",
    "    \n",
    "\n",
    "\n",
    "    \n",
    "############# clean ################\n",
    "for ke in metaProjects.keys():\n",
    "    fh = open(ke+'/'+dataType+analysis+samType+'.targets.clean.'+ke+'.txt', 'wt')\n",
    "    \n",
    "    checkSam = OrderedDict()\n",
    "    samples = []\n",
    "    with open(ke+'/'+dataType+analysis+samType+'.targets.'+ke+'.txt') as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            lst = line.split('\\t')\n",
    "            \n",
    "            if lst[0] == '':\n",
    "                fh.write(line+'\\n')\n",
    "                continue\n",
    "                \n",
    "            else:\n",
    "                samples.append(lst[0])\n",
    "                checkSam[lst[0]] = line\n",
    "                \n",
    "    uniSam = OrderedDict()\n",
    "    \n",
    "    for sam in samples:\n",
    "        for x in string.ascii_uppercase:\n",
    "            xSam = re.sub('[A-Z]$', x, sam)\n",
    "\n",
    "            if xSam in samples:\n",
    "                uniSam[xSam] = 1\n",
    "                break\n",
    "\n",
    "    for sam in sorted(uniSam.keys()):\n",
    "        fh.write(checkSam[sam]+'\\n')\n",
    "    \n",
    "    fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "######################################################\n",
    "############# Extract mature miRNA ###################\n",
    "######################################################\n",
    "\n",
    "mirbase21 = OrderedDict()\n",
    "\n",
    "with open(databases+'miRBase/mature.release21.id.txt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        mirbase21[lst[0]] = lst[1]\n",
    "\n",
    "\n",
    "for prjt in projects.keys():\n",
    "    \n",
    "    if not os.path.isdir(easyTCGA+prjt+'/'+isoform):\n",
    "        continue\n",
    "    \n",
    "    os.chdir(easyTCGA+prjt+'/'+isoform)\n",
    "    files = os.listdir(clean)\n",
    "\n",
    "    for fl in files:\n",
    "        if fl.endswith('.txt') and not fl.startswith('mature'):\n",
    "            fh = open(clean+'mature.counts.'+fl, 'wt')\n",
    "        \n",
    "            mir = OrderedDict()\n",
    "            with open(clean+fl) as f:\n",
    "                for line in f:\n",
    "                    line = line.rstrip()\n",
    "        \n",
    "                    lst = line.split('\\t')\n",
    "        \n",
    "                    if not lst[-1].startswith('mature'):\n",
    "                        continue\n",
    "            \n",
    "                    else:\n",
    "                        mimat = lst[-1].split(',')[-1]\n",
    "            \n",
    "                        if mimat not in mir.keys():\n",
    "                            mir[mimat] = 0\n",
    "                        mir[mimat] += float(lst[2])\n",
    "            \n",
    "            \n",
    "            for ke in mirbase21.keys():\n",
    "                if ke not in mir.keys():\n",
    "                    fh.write(mirbase21[ke]+'\\t0\\n')\n",
    "        \n",
    "                else:\n",
    "                    fh.write(mirbase21[ke]+'\\t'+str(mir[ke])+'\\n')\n",
    "    \n",
    "    \n",
    "            fh.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "####################################################\n",
    "############# Merge mature miRNA ###################\n",
    "####################################################\n",
    "\n",
    "for prjt in projects:\n",
    "    \n",
    "    if not os.path.isdir(easyTCGA+prjt+'/'+isoform):\n",
    "        continue\n",
    "    \n",
    "    samples = []\n",
    "    uuid2sample = OrderedDict()\n",
    "    with open(easyTCGA+'isoform.metadata.cart.2017-02-06.all.basic_information.txt') as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "        \n",
    "            lst = line.split('\\t')\n",
    "        \n",
    "            if lst[-1] == prjt:\n",
    "                samples.append(lst[3])\n",
    "                uuid2sample[lst[1]] = lst[3]\n",
    "                \n",
    "    \n",
    "    uniSam = OrderedDict()\n",
    "    for sam in samples:\n",
    "        for x in string.ascii_uppercase:\n",
    "            xSam = re.sub('[A-Z]$', x, sam)\n",
    "\n",
    "            if xSam in samples:\n",
    "                uniSam[xSam] = 1\n",
    "                break\n",
    "    \n",
    "    \n",
    "    \n",
    "    os.chdir(easyTCGA+prjt+'/'+isoform)\n",
    "    files = sorted(os.listdir(clean))\n",
    "    \n",
    "    genoSamples = []\n",
    "    genotype = OrderedDict()\n",
    "    \n",
    "    for fl in files:\n",
    "        if fl.startswith('mature.counts'):\n",
    "            flName = fl.split('.')[-2]\n",
    "            genoSamples.append(uuid2sample[flName])\n",
    "        \n",
    "            with open(clean+fl, 'rt') as f:\n",
    "                for line in f:\n",
    "                    if re.match('miRNA_ID|sample|Composite', line):\n",
    "                        continue\n",
    "                    line = line.rstrip()\n",
    "                    lst = line.split('\\t')\n",
    "                \n",
    "                    if lst[0] not in genotype.keys():\n",
    "                        genotype[lst[0]] = []\n",
    "                    genotype[lst[0]].append(lst[1])\n",
    "    \n",
    "    \n",
    "    \n",
    "    ##### all samples\n",
    "    fh = open(analysis+'mature_miRNAs_expression.counts.'+prjt+'.txt', 'wt')\n",
    "\n",
    "    fh.write('\\t'+'\\t'.join(genoSamples)+'\\n')\n",
    "    \n",
    "    for ke, val in genotype.items():\n",
    "        fh.write(ke+'\\t'+'\\t'.join(val)+'\\n')\n",
    "\n",
    "    fh.close()\n",
    "    \n",
    "    \n",
    "    ##### clean\n",
    "    fh = open(analysis+'mature_miRNAs_expression.counts.clean.'+prjt+'.txt', 'wt')\n",
    "\n",
    "    samples = sorted(uniSam.keys())\n",
    "\n",
    "    sampleIndex = [genoSamples.index(sam) for sam in samples]                               \n",
    "    fh.write('\\t'+'\\t'.join([genoSamples[i] for i in sampleIndex])+'\\n')\n",
    "\n",
    "    for ke, val in genotype.items():\n",
    "        fh.write(ke+'\\t'+'\\t'.join([val[i] for i in sampleIndex])+'\\n')\n",
    "\n",
    "    fh.close()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merge RNAseq data from project TCGA-BRCA is completed !\n",
      "Clean RNAseq data from project TCGA-BRCA is completed !\n",
      "Merge RNAseq data from project TCGA-GBM is completed !\n",
      "Clean RNAseq data from project TCGA-GBM is completed !\n",
      "Merge RNAseq data from project TCGA-OV is completed !\n",
      "Clean RNAseq data from project TCGA-OV is completed !\n",
      "Merge RNAseq data from project TCGA-LUAD is completed !\n",
      "Clean RNAseq data from project TCGA-LUAD is completed !\n",
      "Merge RNAseq data from project TCGA-UCEC is completed !\n",
      "Clean RNAseq data from project TCGA-UCEC is completed !\n",
      "Merge RNAseq data from project TCGA-KIRC is completed !\n",
      "Clean RNAseq data from project TCGA-KIRC is completed !\n",
      "Merge RNAseq data from project TCGA-HNSC is completed !\n",
      "Clean RNAseq data from project TCGA-HNSC is completed !\n",
      "Merge RNAseq data from project TCGA-LGG is completed !\n",
      "Clean RNAseq data from project TCGA-LGG is completed !\n",
      "Merge RNAseq data from project TCGA-THCA is completed !\n",
      "Clean RNAseq data from project TCGA-THCA is completed !\n",
      "Merge RNAseq data from project TCGA-LUSC is completed !\n",
      "Clean RNAseq data from project TCGA-LUSC is completed !\n",
      "Merge RNAseq data from project TCGA-PRAD is completed !\n",
      "Clean RNAseq data from project TCGA-PRAD is completed !\n",
      "Merge RNAseq data from project TCGA-STAD is completed !\n",
      "Clean RNAseq data from project TCGA-STAD is completed !\n",
      "Merge RNAseq data from project TCGA-SKCM is completed !\n",
      "Clean RNAseq data from project TCGA-SKCM is completed !\n",
      "Merge RNAseq data from project TCGA-COAD is completed !\n",
      "Clean RNAseq data from project TCGA-COAD is completed !\n",
      "Merge RNAseq data from project TCGA-BLCA is completed !\n",
      "Clean RNAseq data from project TCGA-BLCA is completed !\n",
      "Merge RNAseq data from project TCGA-LIHC is completed !\n",
      "Clean RNAseq data from project TCGA-LIHC is completed !\n",
      "Merge RNAseq data from project TCGA-CESC is completed !\n",
      "Clean RNAseq data from project TCGA-CESC is completed !\n",
      "Merge RNAseq data from project TCGA-KIRP is completed !\n",
      "Clean RNAseq data from project TCGA-KIRP is completed !\n",
      "Merge RNAseq data from project TCGA-SARC is completed !\n",
      "Clean RNAseq data from project TCGA-SARC is completed !\n",
      "Merge RNAseq data from project TCGA-LAML is completed !\n",
      "Clean RNAseq data from project TCGA-LAML is completed !\n",
      "Merge RNAseq data from project TCGA-PAAD is completed !\n",
      "Clean RNAseq data from project TCGA-PAAD is completed !\n",
      "Merge RNAseq data from project TCGA-ESCA is completed !\n",
      "Clean RNAseq data from project TCGA-ESCA is completed !\n",
      "Merge RNAseq data from project TCGA-PCPG is completed !\n",
      "Clean RNAseq data from project TCGA-PCPG is completed !\n",
      "Merge RNAseq data from project TCGA-READ is completed !\n",
      "Clean RNAseq data from project TCGA-READ is completed !\n",
      "Merge RNAseq data from project TCGA-TGCT is completed !\n",
      "Clean RNAseq data from project TCGA-TGCT is completed !\n",
      "Merge RNAseq data from project TCGA-THYM is completed !\n",
      "Clean RNAseq data from project TCGA-THYM is completed !\n",
      "Merge RNAseq data from project TCGA-KICH is completed !\n",
      "Clean RNAseq data from project TCGA-KICH is completed !\n",
      "Merge RNAseq data from project TCGA-ACC is completed !\n",
      "Clean RNAseq data from project TCGA-ACC is completed !\n",
      "Merge RNAseq data from project TCGA-MESO is completed !\n",
      "Clean RNAseq data from project TCGA-MESO is completed !\n",
      "Merge RNAseq data from project TCGA-UVM is completed !\n",
      "Clean RNAseq data from project TCGA-UVM is completed !\n",
      "Merge RNAseq data from project TCGA-DLBC is completed !\n",
      "Clean RNAseq data from project TCGA-DLBC is completed !\n",
      "Merge RNAseq data from project TCGA-UCS is completed !\n",
      "Clean RNAseq data from project TCGA-UCS is completed !\n",
      "Merge RNAseq data from project TCGA-CHOL is completed !\n",
      "Clean RNAseq data from project TCGA-CHOL is completed !\n"
     ]
    }
   ],
   "source": [
    "###################################################\n",
    "############# Merge RNAseq data ###################\n",
    "###################################################\n",
    "\n",
    "\n",
    "for prjt in projects.keys():\n",
    "    \n",
    "    if not os.path.isdir(easyTCGA+prjt+'/'+RNAseq):\n",
    "        continue\n",
    "    \n",
    "    samples = []\n",
    "    uuid2sample = OrderedDict()\n",
    "    with open(easyTCGA+'RNAseq.metadata.cart.2017-02-03.all.basic_information.txt') as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "        \n",
    "            lst = line.split('\\t')\n",
    "        \n",
    "            if lst[-1] == prjt:\n",
    "                samples.append(lst[3])\n",
    "                uuid2sample[lst[1]] = lst[3]\n",
    "                \n",
    "    \n",
    "    uniSam = OrderedDict()\n",
    "    for sam in samples:\n",
    "        for x in string.ascii_uppercase:\n",
    "            xSam = re.sub('[A-Z]$', x, sam)\n",
    "\n",
    "            if xSam in samples:\n",
    "                uniSam[xSam] = 1\n",
    "                break\n",
    "    \n",
    "    \n",
    "    \n",
    "    os.chdir(easyTCGA+prjt+'/'+RNAseq)\n",
    "    files = sorted(os.listdir(clean))\n",
    "    \n",
    "    genoSamples = []\n",
    "    genotype = OrderedDict()\n",
    "    \n",
    "    for fl in files:\n",
    "        flName = fl.split('.')[0]\n",
    "        genoSamples.append(uuid2sample[flName])\n",
    "        \n",
    "        with open(clean+fl, 'rt') as f:\n",
    "            for line in f:\n",
    "                if re.match('miRNA_ID|sample|Composite|__', line):\n",
    "                    continue\n",
    "                line = line.rstrip()\n",
    "                lst = line.split('\\t')\n",
    "                \n",
    "                if lst[0] not in genotype.keys():\n",
    "                    genotype[lst[0]] = []\n",
    "                genotype[lst[0]].append(lst[1])\n",
    "    \n",
    "    \n",
    "    \n",
    "    ##### all samples\n",
    "    fh = open(analysis+'RNAseq.raw_counts.'+prjt+'.txt', 'wt')\n",
    "\n",
    "    fh.write('\\t'+'\\t'.join(genoSamples)+'\\n')\n",
    "    \n",
    "    for ke, val in genotype.items():\n",
    "        fh.write(ke+'\\t'+'\\t'.join(val)+'\\n')\n",
    "\n",
    "    fh.close()\n",
    "    \n",
    "    print ('Merge RNAseq data from project %s is completed !' %prjt)\n",
    "    \n",
    "    \n",
    "    ##### clean\n",
    "    fh = open(analysis+'RNAseq.raw_counts.clean.'+prjt+'.txt', 'wt')\n",
    "\n",
    "    samples = sorted(uniSam.keys())\n",
    "\n",
    "    sampleIndex = [genoSamples.index(sam) for sam in samples]                               \n",
    "    fh.write('\\t'+'\\t'.join([genoSamples[i] for i in sampleIndex])+'\\n')\n",
    "\n",
    "    for ke, val in genotype.items():\n",
    "        fh.write(ke+'\\t'+'\\t'.join([val[i] for i in sampleIndex])+'\\n')\n",
    "\n",
    "    fh.close()\n",
    "    \n",
    "    print ('Clean RNAseq data from project %s is completed !' %prjt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merge miRNAseq data from project TCGA-BRCA is completed !\n",
      "Clean miRNAseq data from project TCGA-BRCA is completed !\n",
      "Merge miRNAseq data from project TCGA-GBM is completed !\n",
      "Clean miRNAseq data from project TCGA-GBM is completed !\n",
      "Merge miRNAseq data from project TCGA-OV is completed !\n",
      "Clean miRNAseq data from project TCGA-OV is completed !\n",
      "Merge miRNAseq data from project TCGA-LUAD is completed !\n",
      "Clean miRNAseq data from project TCGA-LUAD is completed !\n",
      "Merge miRNAseq data from project TCGA-UCEC is completed !\n",
      "Clean miRNAseq data from project TCGA-UCEC is completed !\n",
      "Merge miRNAseq data from project TCGA-KIRC is completed !\n",
      "Clean miRNAseq data from project TCGA-KIRC is completed !\n",
      "Merge miRNAseq data from project TCGA-HNSC is completed !\n",
      "Clean miRNAseq data from project TCGA-HNSC is completed !\n",
      "Merge miRNAseq data from project TCGA-LGG is completed !\n",
      "Clean miRNAseq data from project TCGA-LGG is completed !\n",
      "Merge miRNAseq data from project TCGA-THCA is completed !\n",
      "Clean miRNAseq data from project TCGA-THCA is completed !\n",
      "Merge miRNAseq data from project TCGA-LUSC is completed !\n",
      "Clean miRNAseq data from project TCGA-LUSC is completed !\n",
      "Merge miRNAseq data from project TCGA-PRAD is completed !\n",
      "Clean miRNAseq data from project TCGA-PRAD is completed !\n",
      "Merge miRNAseq data from project TCGA-STAD is completed !\n",
      "Clean miRNAseq data from project TCGA-STAD is completed !\n",
      "Merge miRNAseq data from project TCGA-SKCM is completed !\n",
      "Clean miRNAseq data from project TCGA-SKCM is completed !\n",
      "Merge miRNAseq data from project TCGA-COAD is completed !\n",
      "Clean miRNAseq data from project TCGA-COAD is completed !\n",
      "Merge miRNAseq data from project TCGA-BLCA is completed !\n",
      "Clean miRNAseq data from project TCGA-BLCA is completed !\n",
      "Merge miRNAseq data from project TCGA-LIHC is completed !\n",
      "Clean miRNAseq data from project TCGA-LIHC is completed !\n",
      "Merge miRNAseq data from project TCGA-CESC is completed !\n",
      "Clean miRNAseq data from project TCGA-CESC is completed !\n",
      "Merge miRNAseq data from project TCGA-KIRP is completed !\n",
      "Clean miRNAseq data from project TCGA-KIRP is completed !\n",
      "Merge miRNAseq data from project TCGA-SARC is completed !\n",
      "Clean miRNAseq data from project TCGA-SARC is completed !\n",
      "Merge miRNAseq data from project TCGA-LAML is completed !\n",
      "Clean miRNAseq data from project TCGA-LAML is completed !\n",
      "Merge miRNAseq data from project TCGA-PAAD is completed !\n",
      "Clean miRNAseq data from project TCGA-PAAD is completed !\n",
      "Merge miRNAseq data from project TCGA-ESCA is completed !\n",
      "Clean miRNAseq data from project TCGA-ESCA is completed !\n",
      "Merge miRNAseq data from project TCGA-PCPG is completed !\n",
      "Clean miRNAseq data from project TCGA-PCPG is completed !\n",
      "Merge miRNAseq data from project TCGA-READ is completed !\n",
      "Clean miRNAseq data from project TCGA-READ is completed !\n",
      "Merge miRNAseq data from project TCGA-TGCT is completed !\n",
      "Clean miRNAseq data from project TCGA-TGCT is completed !\n",
      "Merge miRNAseq data from project TCGA-THYM is completed !\n",
      "Clean miRNAseq data from project TCGA-THYM is completed !\n",
      "Merge miRNAseq data from project TCGA-KICH is completed !\n",
      "Clean miRNAseq data from project TCGA-KICH is completed !\n",
      "Merge miRNAseq data from project TCGA-ACC is completed !\n",
      "Clean miRNAseq data from project TCGA-ACC is completed !\n",
      "Merge miRNAseq data from project TCGA-MESO is completed !\n",
      "Clean miRNAseq data from project TCGA-MESO is completed !\n",
      "Merge miRNAseq data from project TCGA-UVM is completed !\n",
      "Clean miRNAseq data from project TCGA-UVM is completed !\n",
      "Merge miRNAseq data from project TCGA-DLBC is completed !\n",
      "Clean miRNAseq data from project TCGA-DLBC is completed !\n",
      "Merge miRNAseq data from project TCGA-UCS is completed !\n",
      "Clean miRNAseq data from project TCGA-UCS is completed !\n",
      "Merge miRNAseq data from project TCGA-CHOL is completed !\n",
      "Clean miRNAseq data from project TCGA-CHOL is completed !\n"
     ]
    }
   ],
   "source": [
    "###################################################\n",
    "############# Merge miRNA data ####################\n",
    "###################################################\n",
    "\n",
    "\n",
    "for prjt in projects.keys():\n",
    "    \n",
    "    if not os.path.isdir(easyTCGA+prjt+'/'+miRNA):\n",
    "        continue\n",
    "    \n",
    "    samples = []\n",
    "    uuid2sample = OrderedDict()\n",
    "    with open(easyTCGA+'miRseq.metadata.cart.2017-02-06.all.basic_information.txt') as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "        \n",
    "            lst = line.split('\\t')\n",
    "        \n",
    "            if lst[-1] == prjt:\n",
    "                samples.append(lst[3])\n",
    "                uuid2sample[lst[1]] = lst[3]\n",
    "                \n",
    "    \n",
    "    uniSam = OrderedDict()\n",
    "    for sam in samples:\n",
    "        for x in string.ascii_uppercase:\n",
    "            xSam = re.sub('[A-Z]$', x, sam)\n",
    "\n",
    "            if xSam in samples:\n",
    "                uniSam[xSam] = 1\n",
    "                break\n",
    "    \n",
    "    \n",
    "    \n",
    "    os.chdir(easyTCGA+prjt+'/'+miRNA)\n",
    "    files = sorted(os.listdir(clean))\n",
    "    \n",
    "    genoSamples = []\n",
    "    genotype = OrderedDict()\n",
    "    \n",
    "    for fl in files:\n",
    "        flName = fl.split('.')[0]\n",
    "        #if flName not in uuid2sample.keys():\n",
    "        #    print (flName)\n",
    "        #    continue\n",
    "        genoSamples.append(uuid2sample[flName])\n",
    "        \n",
    "        with open(clean+fl, 'rt') as f:\n",
    "            for line in f:\n",
    "                if re.match('miRNA_ID|sample|Composite|__', line):\n",
    "                    continue\n",
    "                line = line.rstrip()\n",
    "                lst = line.split('\\t')\n",
    "                \n",
    "                if lst[0] not in genotype.keys():\n",
    "                    genotype[lst[0]] = []\n",
    "                genotype[lst[0]].append(lst[1])  ### counts:lst[1], rpm:lst[2]\n",
    "    \n",
    "    \n",
    "    \n",
    "    ##### all samples\n",
    "    fh = open(analysis+'miRNAseq.raw_counts.'+prjt+'.txt', 'wt')\n",
    "\n",
    "    fh.write('\\t'+'\\t'.join(genoSamples)+'\\n')\n",
    "    \n",
    "    for ke, val in genotype.items():\n",
    "        fh.write(ke+'\\t'+'\\t'.join(val)+'\\n')\n",
    "\n",
    "    fh.close()\n",
    "    \n",
    "    print ('Merge miRNAseq data from project %s is completed !' %prjt)\n",
    "    \n",
    "    \n",
    "    ##### clean\n",
    "    fh = open(analysis+'miRNAseq.raw_counts.clean.'+prjt+'.txt', 'wt')\n",
    "\n",
    "    samples = sorted(uniSam.keys())\n",
    "\n",
    "    sampleIndex = [genoSamples.index(sam) for sam in samples]                               \n",
    "    fh.write('\\t'+'\\t'.join([genoSamples[i] for i in sampleIndex])+'\\n')\n",
    "\n",
    "    for ke, val in genotype.items():\n",
    "        fh.write(ke+'\\t'+'\\t'.join([val[i] for i in sampleIndex])+'\\n')\n",
    "\n",
    "    fh.close()\n",
    "    \n",
    "    print ('Clean miRNAseq data from project %s is completed !' %prjt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "###################################################\n",
    "############# Merge methylation data ##############\n",
    "###################################################\n",
    "\n",
    "\n",
    "for prjt in projects.keys():\n",
    "    \n",
    "    if not os.path.isdir(easyTCGA+prjt+'/'+methylation450):\n",
    "        continue\n",
    "    \n",
    "    samples = []\n",
    "    uuid2sample = OrderedDict()\n",
    "    with open(easyTCGA+'miRseq.metadata.cart.2017-02-06.all.basic_information.txt') as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "        \n",
    "            lst = line.split('\\t')\n",
    "        \n",
    "            if lst[-1] == prjt:\n",
    "                samples.append(lst[3])\n",
    "                uuid2sample[lst[1]] = lst[3]\n",
    "                \n",
    "    \n",
    "    uniSam = OrderedDict()\n",
    "    for sam in samples:\n",
    "        for x in string.ascii_uppercase:\n",
    "            xSam = re.sub('[A-Z]$', x, sam)\n",
    "\n",
    "            if xSam in samples:\n",
    "                uniSam[xSam] = 1\n",
    "                break\n",
    "    \n",
    "    \n",
    "    \n",
    "    os.chdir(easyTCGA+prjt+'/'+methylation450)\n",
    "    files = sorted(os.listdir(clean))\n",
    "    \n",
    "    genoSamples = []\n",
    "    genotype = OrderedDict()\n",
    "    \n",
    "    for fl in files:\n",
    "        flName = fl.split('.')[0]\n",
    "        #if flName not in uuid2sample.keys():\n",
    "        #    print (flName)\n",
    "        #    continue\n",
    "        genoSamples.append(uuid2sample[flName])\n",
    "        \n",
    "        with open(clean+fl, 'rt') as f:\n",
    "            for line in f:\n",
    "                if re.match('miRNA_ID|sample|Composite|__', line):\n",
    "                    continue\n",
    "                line = line.rstrip()\n",
    "                lst = line.split('\\t')\n",
    "                \n",
    "                if lst[0] not in genotype.keys():\n",
    "                    genotype[lst[0]] = []\n",
    "                genotype[lst[0]].append(lst[1])  ### counts:lst[1], rpm:lst[2]\n",
    "    \n",
    "    \n",
    "    \n",
    "    ##### all samples\n",
    "    fh = open(analysis+'methylation450.beta_value.'+prjt+'.txt', 'wt')\n",
    "\n",
    "    fh.write('\\t'+'\\t'.join(genoSamples)+'\\n')\n",
    "    \n",
    "    for ke, val in genotype.items():\n",
    "        fh.write(ke+'\\t'+'\\t'.join(val)+'\\n')\n",
    "\n",
    "    fh.close()\n",
    "    \n",
    "    print ('Merge methylation450 data from project %s is completed !' %prjt)\n",
    "    \n",
    "    \n",
    "    ##### clean\n",
    "    fh = open(analysis+'methylation450.beta_value.clean.'+prjt+'.txt', 'wt')\n",
    "\n",
    "    samples = sorted(uniSam.keys())\n",
    "\n",
    "    sampleIndex = [genoSamples.index(sam) for sam in samples]                               \n",
    "    fh.write('\\t'+'\\t'.join([genoSamples[i] for i in sampleIndex])+'\\n')\n",
    "\n",
    "    for ke, val in genotype.items():\n",
    "        fh.write(ke+'\\t'+'\\t'.join([val[i] for i in sampleIndex])+'\\n')\n",
    "\n",
    "    fh.close()\n",
    "    \n",
    "    print ('Clean methylation450 data from project %s is completed !' %prjt)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metadata file name (no .json extension):\n",
      "clinical.metadata.cart.2017-02-05.all\n",
      "\n",
      "################################################################\n",
      "Number of Clinical_Supplement samples in project TCGA-LUAD: 522\n",
      "Number of Clinical_Supplement samples in project TCGA-GBM: 596\n",
      "Number of Clinical_Supplement samples in project TCGA-LGG: 515\n",
      "Number of Clinical_Supplement samples in project TCGA-OV: 587\n",
      "Number of Clinical_Supplement samples in project TCGA-LUSC: 504\n",
      "Number of Clinical_Supplement samples in project TCGA-PRAD: 500\n",
      "Number of Clinical_Supplement samples in project TCGA-BRCA: 1097\n",
      "Number of Clinical_Supplement samples in project TCGA-BLCA: 412\n",
      "Number of Clinical_Supplement samples in project TCGA-PCPG: 179\n",
      "Number of Clinical_Supplement samples in project TCGA-MESO: 87\n",
      "Number of Clinical_Supplement samples in project TCGA-THYM: 124\n",
      "Number of Clinical_Supplement samples in project TCGA-LIHC: 377\n",
      "Number of Clinical_Supplement samples in project TCGA-SARC: 261\n",
      "Number of Clinical_Supplement samples in project TCGA-PAAD: 185\n",
      "Number of Clinical_Supplement samples in project TCGA-KIRC: 537\n",
      "Number of Clinical_Supplement samples in project TCGA-STAD: 443\n",
      "Number of Clinical_Supplement samples in project TCGA-CESC: 307\n",
      "Number of Clinical_Supplement samples in project TCGA-ACC: 92\n",
      "Number of Clinical_Supplement samples in project TCGA-SKCM: 470\n",
      "Number of Clinical_Supplement samples in project TCGA-ESCA: 185\n",
      "Number of Clinical_Supplement samples in project TCGA-UCEC: 548\n",
      "Number of Clinical_Supplement samples in project TCGA-HNSC: 528\n",
      "Number of Clinical_Supplement samples in project TCGA-THCA: 507\n",
      "Number of Clinical_Supplement samples in project TCGA-KIRP: 291\n",
      "Number of Clinical_Supplement samples in project TCGA-READ: 170\n",
      "Number of Clinical_Supplement samples in project TCGA-DLBC: 48\n",
      "Number of Clinical_Supplement samples in project TCGA-TGCT: 134\n",
      "Number of Clinical_Supplement samples in project TCGA-COAD: 459\n",
      "Number of Clinical_Supplement samples in project TCGA-LAML: 200\n",
      "Number of Clinical_Supplement samples in project TCGA-UVM: 80\n",
      "Number of Clinical_Supplement samples in project TCGA-KICH: 113\n",
      "Number of Clinical_Supplement samples in project TCGA-CHOL: 45\n",
      "Number of Clinical_Supplement samples in project TCGA-UCS: 57\n"
     ]
    }
   ],
   "source": [
    "#######################################\n",
    "#### Step1. Parse Simple Metadata #####\n",
    "#######################################\n",
    "\n",
    "\n",
    "def parseAnnotations(metadatai):\n",
    "    annotationInfo = ['notes', 'category', 'status']\n",
    "    \n",
    "    if 'annotations' not in metadatai.keys():\n",
    "        notes, category, status = None, None, None\n",
    "        annotations = [notes, category, status]\n",
    "    else:\n",
    "        annoNum = len(metadatai['annotations'])\n",
    "        notes, category, status = [None]*annoNum, [None]*annoNum, [None]*annoNum\n",
    "        for j in range(annoNum):\n",
    "            [notes[j], category[j], status[j]] = [metadatai['annotations'][j][anno] for anno in annotationInfo]\n",
    "        annotations = [';'.join(map(str,notes)), ';'.join(map(str,category)), ';'.join(map(str,status))]\n",
    "        annotations = [re.sub('\\r+|\\n+','',anno) for anno in annotations]\n",
    "    return annotations\n",
    "\n",
    "\n",
    "def tcgaMetaParser(metaFile, genoDataType):\n",
    "    \n",
    "    sampleInfo = ['submitter_id', 'tissue_type', 'sample_type', 'tumor_code_id', 'sample_type_id']\n",
    "    basicDemographic = ['race', 'gender', 'ethnicity']\n",
    "    basicDiagnoses = ['tumor_stage','tumor_grade','age_at_diagnosis','days_to_death','days_to_last_known_disease_status',\n",
    "                      'days_to_last_follow_up','vital_status']\n",
    "    annotationInfo = ['notes', 'category', 'status']\n",
    "    \n",
    "    fhTmp = open(metaFile+'.basic_information.tmp', 'wt')\n",
    "\n",
    "    \n",
    "    with open(metaFile+'.json', 'rt') as f:\n",
    "        metadata = json.load(f)\n",
    "        \n",
    "    for i in range(len(metadata)):\n",
    " \n",
    "        fileName = metadata[i]['file_name']\n",
    "        fileId = metadata[i]['file_id']\n",
    "        dataType = metadata[i]['data_type'].replace(' ', '_')\n",
    "\n",
    "        annotations = parseAnnotations(metadata[i])\n",
    "        \n",
    "        for k in metadata[i].keys():\n",
    "            if k == 'cases':\n",
    "                caseId = metadata[i][k][0]['case_id']\n",
    "                projectId = metadata[i][k][0]['project']['project_id']\n",
    "                \n",
    "                \n",
    "                if genoDataType in [7,8]:\n",
    "                    [submitterId, tissueType, sampleType, tumorCodeId, sampleTypeId] = [None]*5\n",
    "                    submitterId = fileName.split('.')[-2]\n",
    "                else:\n",
    "                    [submitterId, tissueType, sampleType, tumorCodeId, sampleTypeId] = \\\n",
    "                    [metadata[i][k][0]['samples'][0][sam] for sam in sampleInfo]\n",
    "\n",
    "                if 'diagnoses' not in metadata[i][k][0].keys():\n",
    "                    numBlank = len(basicDemographic)+len(basicDiagnoses)\n",
    "                    \n",
    "                    fhTmp.write('\\t'.join(map(str,[fileName,fileId,caseId,submitterId,sampleType,sampleTypeId]\\\n",
    "                                                +[None]*numBlank+annotations+[dataType,projectId]))+'\\n')\n",
    "                    \n",
    "                else:\n",
    "                    basicDemo = [race, gender, ethnicity]\\\n",
    "                              = [metadata[i][k][0]['demographic'][demo] for demo in basicDemographic]\n",
    "                \n",
    "                    basicDiag = [tumorStage,tumorGrade,ageAtDiagnosis,daysToDeath,daysToLastKnownStatus,daysToLastFollowUp,\n",
    "                             vitalStatus]=[metadata[i][k][0]['diagnoses'][0][diag] for diag in basicDiagnoses]\n",
    "                \n",
    "\n",
    "                    fhTmp.write('\\t'.join(map(str, [fileName,fileId,caseId,submitterId,sampleType,sampleTypeId]\\\n",
    "                                    +basicDiag+basicDemo+annotations+[dataType,projectId]))+'\\n')\n",
    "        \n",
    "    fhTmp.close()\n",
    "    \n",
    "    fh = open(metaFile+'.basic_information.txt', 'wt')\n",
    "    fh.write('\\t'.join(map(str, ['file_name','file_uuid','case_id','submitter_id','sample_type','sample_type_id']\\\n",
    "            +basicDiagnoses+basicDemographic+annotationInfo+['data_type','project_id']))+'\\n')\n",
    "\n",
    "    newFile = OrderedDict()\n",
    "    with open(metaFile+'.basic_information.tmp', 'rt') as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            lst = line.split('\\t')\n",
    "            if lst[-1] not in newFile.keys():\n",
    "                newFile[lst[-1]] = OrderedDict()\n",
    "            newFile[lst[-1]][lst[1]] = line\n",
    "    \n",
    "    print ()\n",
    "    print ('################################################################')\n",
    "    \n",
    "    for ke,val in newFile.items():\n",
    "        print ('Number of %s samples in project %s: %d' %(dataFolders[genoDataType],ke,len(val)))\n",
    "        \n",
    "        for subk, subv in val.items():\n",
    "            subv = re.sub(r'not reported|None', 'NA', subv)\n",
    "            fh.write(subv+'\\n')\n",
    "    \n",
    "    fh.close()\n",
    "\n",
    "    os.system('rm '+metaFile+'.basic_information.tmp')\n",
    "\n",
    "#########################################################################\n",
    "metaFile = input('Metadata file name (no .json extension):\\n')\n",
    "tcgaMetaParser(metaFile,genoDataType)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metadata file name (no .json extension):\n",
      "clinical.metadata.cart.2017-02-05.all\n",
      "Number of projects: 33\n",
      "Total Number of samples: 11160\n",
      "#################################################\n",
      "Number of samples in project TCGA-LUAD: 522\n",
      "Number of samples in project TCGA-GBM: 596\n",
      "Number of samples in project TCGA-LGG: 515\n",
      "Number of samples in project TCGA-OV: 587\n",
      "Number of samples in project TCGA-LUSC: 504\n",
      "Number of samples in project TCGA-PRAD: 500\n",
      "Number of samples in project TCGA-BRCA: 1097\n",
      "Number of samples in project TCGA-BLCA: 412\n",
      "Number of samples in project TCGA-PCPG: 179\n",
      "Number of samples in project TCGA-MESO: 87\n",
      "Number of samples in project TCGA-THYM: 124\n",
      "Number of samples in project TCGA-LIHC: 377\n",
      "Number of samples in project TCGA-SARC: 261\n",
      "Number of samples in project TCGA-PAAD: 185\n",
      "Number of samples in project TCGA-KIRC: 537\n",
      "Number of samples in project TCGA-STAD: 443\n",
      "Number of samples in project TCGA-CESC: 307\n",
      "Number of samples in project TCGA-ACC: 92\n",
      "Number of samples in project TCGA-SKCM: 470\n",
      "Number of samples in project TCGA-ESCA: 185\n",
      "Number of samples in project TCGA-UCEC: 548\n",
      "Number of samples in project TCGA-HNSC: 528\n",
      "Number of samples in project TCGA-THCA: 507\n",
      "Number of samples in project TCGA-KIRP: 291\n",
      "Number of samples in project TCGA-READ: 170\n",
      "Number of samples in project TCGA-DLBC: 48\n",
      "Number of samples in project TCGA-TGCT: 134\n",
      "Number of samples in project TCGA-COAD: 459\n",
      "Number of samples in project TCGA-LAML: 200\n",
      "Number of samples in project TCGA-UVM: 80\n",
      "Number of samples in project TCGA-KICH: 113\n",
      "Number of samples in project TCGA-CHOL: 45\n",
      "Number of samples in project TCGA-UCS: 57\n"
     ]
    }
   ],
   "source": [
    "####################################################\n",
    "######### Step2. Summarize metadata files ##########\n",
    "####################################################\n",
    "\n",
    "try:\n",
    "    metaFile\n",
    "except NameError:\n",
    "    metaFile = input('Metadata file name (no .json extension):\\n')\n",
    "\n",
    "# del metaFile ### can be used to remove a variable\n",
    "\n",
    "metaProjects = OrderedDict()\n",
    "uuid = OrderedDict()\n",
    "samples = OrderedDict()\n",
    "traits = OrderedDict()\n",
    "\n",
    "with open(easyTCGA+metaFile+'.basic_information.txt', 'rt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        if line.startswith('file_name'):\n",
    "            traitName = lst[4:16]\n",
    "            continue\n",
    "            \n",
    "        elif lst[-1] not in metaProjects.keys():\n",
    "            metaProjects[lst[-1]] = 0\n",
    "            uuid[lst[-1]] = OrderedDict()\n",
    "            traits[lst[-1]] = OrderedDict()\n",
    "            samples[lst[-1]] = []\n",
    "            \n",
    "        metaProjects[lst[-1]] += 1\n",
    "        uuid[lst[-1]][lst[1]] = lst[3]\n",
    "        samples[lst[-1]].append(lst[3])\n",
    "        \n",
    "        for i in range(len(traitName)):\n",
    "            trt = traitName[i]\n",
    "            if trt not in traits[lst[-1]].keys():\n",
    "                traits[lst[-1]][trt] = []\n",
    "            traits[lst[-1]][trt].append(lst[i+4].replace(' ',''))\n",
    "            \n",
    "            \n",
    "print ('Number of projects: %d' %len(metaProjects))\n",
    "print ('Total Number of samples: %d' %sum([val for val in metaProjects.values()]))\n",
    "print ('#################################################')\n",
    "\n",
    "for ke, val in metaProjects.items():\n",
    "    print ('Number of samples in project %s: %d' %(ke,val))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#############################################\n",
    "##### Step3. Create required directories ####\n",
    "#############################################\n",
    "\n",
    "### if exits, do nothing\n",
    "dataType = ['Gene_Expression_Quantification',\n",
    "            'miRNA_Expression_Quantification',\n",
    "            'Isoform_Expression_Quantification',\n",
    "            'Methylation_Beta_Value27',\n",
    "            'Methylation_Beta_Value450',\n",
    "            'Copy_Number_Segment',\n",
    "            'Masked_Copy_Number_Segment',\n",
    "            'Clinical_Supplement',\n",
    "            'Biospecimen_Supplement',\n",
    "            'Raw Simple Somatic Mutation',\n",
    "            'Masked_Somatic_Mutation']\n",
    "\n",
    "genoFile = ['RNAseq.counts',\n",
    "            'miRseq.counts',\n",
    "            'isoform.counts',\n",
    "            'methylation.beta_value27',\n",
    "            'methylation.beta_value450',\n",
    "            None,\n",
    "            None,\n",
    "            'Merged_clinical_data.',\n",
    "            None,\n",
    "            None]\n",
    "\n",
    "dataFolders = OrderedDict()\n",
    "for i in range(1,11):\n",
    "    dataFolders[i] = dataType[i-1]\n",
    "    \n",
    "genoDataNames = OrderedDict()\n",
    "for i in range(1,11):\n",
    "    genoDataNames[i] = genoFile[i-1]\n",
    "    \n",
    "newFolders = ' '.join([easyTCGA+prjt+'/'+dataFolders[genoDataType]+'/'+fd for prjt in metaProjects \n",
    "                       for fd in ['Downloads','Clean','Analysis']])\n",
    "os.system('mkdir -p '+newFolders)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TCGA-LUAD\tStart Downloading ... (2017-02-05 13:57:17)\n",
      "TCGA-LUAD\tDownloading completed ! (2017-02-05 14:24:50)\n",
      "Number of downloaded Clinical_Supplement samples: 522\n",
      "TCGA-LUAD\tStart Cleaning ...\n",
      "TCGA-LUAD\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 1044\n",
      "################################################\n",
      "TCGA-GBM\tStart Downloading ... (2017-02-05 14:24:52)\n",
      "TCGA-GBM\tDownloading completed ! (2017-02-05 14:48:19)\n",
      "Number of downloaded Clinical_Supplement samples: 596\n",
      "TCGA-GBM\tStart Cleaning ...\n",
      "TCGA-GBM\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 596\n",
      "################################################\n",
      "TCGA-LGG\tStart Downloading ... (2017-02-05 14:48:21)\n",
      "TCGA-LGG\tDownloading completed ! (2017-02-05 15:27:32)\n",
      "Number of downloaded Clinical_Supplement samples: 515\n",
      "TCGA-LGG\tStart Cleaning ...\n",
      "TCGA-LGG\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 515\n",
      "################################################\n",
      "TCGA-OV\tStart Downloading ... (2017-02-05 15:27:34)\n",
      "TCGA-OV\tDownloading completed ! (2017-02-05 16:18:50)\n",
      "Number of downloaded Clinical_Supplement samples: 587\n",
      "TCGA-OV\tStart Cleaning ...\n",
      "TCGA-OV\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 587\n",
      "################################################\n",
      "TCGA-LUSC\tStart Downloading ... (2017-02-05 16:18:53)\n",
      "TCGA-LUSC\tDownloading completed ! (2017-02-05 17:00:16)\n",
      "Number of downloaded Clinical_Supplement samples: 504\n",
      "TCGA-LUSC\tStart Cleaning ...\n",
      "TCGA-LUSC\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 504\n",
      "################################################\n",
      "TCGA-PRAD\tStart Downloading ... (2017-02-05 17:00:19)\n",
      "TCGA-PRAD\tDownloading completed ! (2017-02-05 17:45:28)\n",
      "Number of downloaded Clinical_Supplement samples: 500\n",
      "TCGA-PRAD\tStart Cleaning ...\n",
      "TCGA-PRAD\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 500\n",
      "################################################\n",
      "TCGA-BRCA\tStart Downloading ... (2017-02-05 17:45:30)\n",
      "TCGA-BRCA\tDownloading completed ! (2017-02-05 19:13:49)\n",
      "Number of downloaded Clinical_Supplement samples: 1097\n",
      "TCGA-BRCA\tStart Cleaning ...\n",
      "TCGA-BRCA\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 1097\n",
      "################################################\n",
      "TCGA-BLCA\tStart Downloading ... (2017-02-05 19:13:56)\n",
      "TCGA-BLCA\tDownloading completed ! (2017-02-05 19:47:32)\n",
      "Number of downloaded Clinical_Supplement samples: 412\n",
      "TCGA-BLCA\tStart Cleaning ...\n",
      "TCGA-BLCA\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 412\n",
      "################################################\n",
      "TCGA-PCPG\tStart Downloading ... (2017-02-05 19:47:33)\n",
      "TCGA-PCPG\tDownloading completed ! (2017-02-05 20:02:12)\n",
      "Number of downloaded Clinical_Supplement samples: 179\n",
      "TCGA-PCPG\tStart Cleaning ...\n",
      "TCGA-PCPG\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 179\n",
      "################################################\n",
      "TCGA-MESO\tStart Downloading ... (2017-02-05 20:02:14)\n",
      "TCGA-MESO\tDownloading completed ! (2017-02-05 20:07:44)\n",
      "Number of downloaded Clinical_Supplement samples: 87\n",
      "TCGA-MESO\tStart Cleaning ...\n",
      "TCGA-MESO\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 87\n",
      "################################################\n",
      "TCGA-THYM\tStart Downloading ... (2017-02-05 20:07:44)\n",
      "TCGA-THYM\tDownloading completed ! (2017-02-05 20:16:52)\n",
      "Number of downloaded Clinical_Supplement samples: 124\n",
      "TCGA-THYM\tStart Cleaning ...\n",
      "TCGA-THYM\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 124\n",
      "################################################\n",
      "TCGA-LIHC\tStart Downloading ... (2017-02-05 20:16:53)\n",
      "TCGA-LIHC\tDownloading completed ! (2017-02-05 20:44:08)\n",
      "Number of downloaded Clinical_Supplement samples: 377\n",
      "TCGA-LIHC\tStart Cleaning ...\n",
      "TCGA-LIHC\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 377\n",
      "################################################\n",
      "TCGA-SARC\tStart Downloading ... (2017-02-05 20:44:10)\n",
      "TCGA-SARC\tDownloading completed ! (2017-02-05 21:03:57)\n",
      "Number of downloaded Clinical_Supplement samples: 261\n",
      "TCGA-SARC\tStart Cleaning ...\n",
      "TCGA-SARC\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 261\n",
      "################################################\n",
      "TCGA-PAAD\tStart Downloading ... (2017-02-05 21:03:58)\n",
      "TCGA-PAAD\tDownloading completed ! (2017-02-05 21:17:36)\n",
      "Number of downloaded Clinical_Supplement samples: 185\n",
      "TCGA-PAAD\tStart Cleaning ...\n",
      "TCGA-PAAD\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 185\n",
      "################################################\n",
      "TCGA-KIRC\tStart Downloading ... (2017-02-05 21:17:36)\n",
      "TCGA-KIRC\tDownloading completed ! (2017-02-05 22:02:28)\n",
      "Number of downloaded Clinical_Supplement samples: 537\n",
      "TCGA-KIRC\tStart Cleaning ...\n",
      "TCGA-KIRC\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 537\n",
      "################################################\n",
      "TCGA-STAD\tStart Downloading ... (2017-02-05 22:02:30)\n",
      "TCGA-STAD\tDownloading completed ! (2017-02-05 22:33:18)\n",
      "Number of downloaded Clinical_Supplement samples: 443\n",
      "TCGA-STAD\tStart Cleaning ...\n",
      "TCGA-STAD\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 443\n",
      "################################################\n",
      "TCGA-CESC\tStart Downloading ... (2017-02-05 22:33:20)\n",
      "TCGA-CESC\tDownloading completed ! (2017-02-05 23:03:05)\n",
      "Number of downloaded Clinical_Supplement samples: 307\n",
      "TCGA-CESC\tStart Cleaning ...\n",
      "TCGA-CESC\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 307\n",
      "################################################\n",
      "TCGA-ACC\tStart Downloading ... (2017-02-05 23:03:06)\n",
      "TCGA-ACC\tDownloading completed ! (2017-02-05 23:12:15)\n",
      "Number of downloaded Clinical_Supplement samples: 92\n",
      "TCGA-ACC\tStart Cleaning ...\n",
      "TCGA-ACC\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 92\n",
      "################################################\n",
      "TCGA-SKCM\tStart Downloading ... (2017-02-05 23:12:15)\n",
      "TCGA-SKCM\tDownloading completed ! (2017-02-05 23:55:00)\n",
      "Number of downloaded Clinical_Supplement samples: 470\n",
      "TCGA-SKCM\tStart Cleaning ...\n",
      "TCGA-SKCM\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 470\n",
      "################################################\n",
      "TCGA-ESCA\tStart Downloading ... (2017-02-05 23:55:02)\n",
      "TCGA-ESCA\tDownloading completed ! (2017-02-06 00:07:18)\n",
      "Number of downloaded Clinical_Supplement samples: 185\n",
      "TCGA-ESCA\tStart Cleaning ...\n",
      "TCGA-ESCA\tCleaning completed !\n",
      "Number of cleaned Clinical_Supplement samples: 185\n",
      "################################################\n",
      "TCGA-UCEC\tStart Downloading ... (2017-02-06 00:07:18)\n"
     ]
    }
   ],
   "source": [
    "#########################################\n",
    "#### Step4. Download and clean data #####\n",
    "#########################################\n",
    "\n",
    "\n",
    "def downloadTCGA(project):\n",
    "    \n",
    "    os.chdir(easyTCGA+'/'.join([project,dataFolders[genoDataType],downloads]))\n",
    "   \n",
    "    print ('##############################################################')\n",
    "    \n",
    "    print (project+'\\tStart Downloading ... (%s)' %datetime.now().strftime('%Y-%m-%d %H:%M:%S'))\n",
    "    fileUUID = ' '.join(uuid[project])\n",
    "    os.system('gdc-client download '+fileUUID)\n",
    "    print (project+'\\tDownloading completed ! (%s)' %datetime.now().strftime('%Y-%m-%d %H:%M:%S'))\n",
    "    #numSamples = len([fd for fd in os.listdir('.') if os.path.isdir(fd)]) ###count number of folders\n",
    "    print ('Number of downloaded %s samples: %d' %(dataFolders[genoDataType],len(os.listdir('.'))))\n",
    "\n",
    "### all file names would be changed to the corresponding folder names (uuids)\n",
    "def cleanTCGA(project):\n",
    "    os.chdir(easyTCGA+project+'/'+dataFolders[genoDataType])\n",
    "    \n",
    "    print (project+'\\tStart Cleaning ...')\n",
    "    \n",
    "    folders = os.listdir(downloads)\n",
    "    if genoDataType == 1:\n",
    "        for fd in folders:\n",
    "            if os.path.isdir(downloads+fd):\n",
    "                os.system('cp '+downloads+fd+'/*.gz '+clean+fd+'.gz')\n",
    "        os.system('gunzip '+clean+'*.gz')\n",
    "    \n",
    "    elif genoDataType in [2,3]:\n",
    "        for fd in folders:\n",
    "            if os.path.isdir(downloads+fd):\n",
    "                os.system('cp '+downloads+fd+'/*.quantification.txt '+clean+fd+'.txt')\n",
    "             \n",
    "    elif genoDataType in [4,5]:\n",
    "        for fd in folders:\n",
    "            if os.path.isdir(downloads+fd):\n",
    "                os.system('cp '+downloads+fd+'/*.gdc_hg38.txt '+clean+fd+'.txt')\n",
    "\n",
    "    if genoDataType in [8,9]:\n",
    "        for fd in folders:\n",
    "            if os.path.isdir(downloads+fd):\n",
    "                os.system('cp '+downloads+fd+'/*.xml '+clean+fd+'.xml')\n",
    "    \n",
    "    print (project+'\\tCleaning completed !')\n",
    "    print ('Number of cleaned %s samples: %d' %(dataFolders[genoDataType],len(os.listdir(clean))))\n",
    "    \n",
    "                \n",
    "        \n",
    "############################################################\n",
    "for project in metaProjects.keys():\n",
    "    downloadTCGA(project)\n",
    "    cleanTCGA(project)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merge miRNA_Expression_Quantification data from project TCGA-PRAD is completed !\n",
      "Merge miRNA_Expression_Quantification data from project TCGA-BLCA is completed !\n",
      "Merge miRNA_Expression_Quantification data from project TCGA-LGG is completed !\n"
     ]
    }
   ],
   "source": [
    "##########################################################\n",
    "#### Step5. Merge Genotyping Data Files into One File ####\n",
    "#########################################################\n",
    "\n",
    "def mergeGenotype(project):\n",
    "    os.chdir(easyTCGA+project+'/'+dataFolders[genoDataType])\n",
    "    \n",
    "    files = sorted(os.listdir(clean))\n",
    "    genoSamples = []\n",
    "    genotype = OrderedDict()\n",
    "    \n",
    "    for fl in files:\n",
    "        flName = fl.split('.')[0]\n",
    "        genoSamples.append(uuid[project][flName])\n",
    "        \n",
    "        with open(clean+fl, 'rt') as f:\n",
    "            for line in f:\n",
    "                if re.match('miRNA_ID|sample|Composite', line):\n",
    "                    continue\n",
    "                line = line.rstrip()\n",
    "                lst = line.split('\\t')\n",
    "                \n",
    "                if lst[0] not in genotype.keys():\n",
    "                    genotype[lst[0]] = []\n",
    "                genotype[lst[0]].append(lst[1])\n",
    "                \n",
    "    fh = open(analysis+genoDataNames[genoDataType]+'_only.'+project+'.txt', 'wt')\n",
    "    \n",
    "    sampleIndex = [genoSamples.index(sam) for sam in samples[project]]\n",
    "                                \n",
    "    fh.write('\\t'+'\\t'.join([genoSamples[i] for i in sampleIndex])+'\\n')\n",
    "    \n",
    "    for ke, val in genotype.items():\n",
    "        fh.write(ke+'\\t'+'\\t'.join([val[i] for i in sampleIndex])+'\\n')\n",
    "\n",
    "    fh.close()\n",
    "    \n",
    "    print ('Merge %s data from project %s is completed !' %(dataFolders[genoDataType], project))\n",
    "    \n",
    "    \n",
    "    \n",
    "##################################################\n",
    "for project in metaProjects.keys():\n",
    "        mergeGenotype(project)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Prepare miRNA_Expression_Quantification data from project TCGA-PRAD is completed !\n",
      "Prepare miRNA_Expression_Quantification data from project TCGA-BLCA is completed !\n",
      "Prepare miRNA_Expression_Quantification data from project TCGA-LGG is completed !\n"
     ]
    }
   ],
   "source": [
    "################################################################\n",
    "########## Step6. Prepare input for downstream analysis ########\n",
    "################################################################\n",
    "\n",
    "def limmaInput(project):\n",
    "    os.chdir(easyTCGA+project+'/'+dataFolders[genoDataType])\n",
    "    \n",
    "    fh = open(analysis+genoDataNames[genoDataType]+'_and_traits.'+project+'.txt', 'wt')\n",
    "\n",
    "    with open(analysis+genoDataNames[genoDataType]+'_only.'+project+'.txt', 'rt') as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            lst = line.split('\\t')\n",
    "            if lst[0] == '':\n",
    "                fh.write(line+'\\n')\n",
    "                \n",
    "                for ke, val in traits[project].items():\n",
    "                    fh.write(ke+'\\t'+'\\t'.join(map(str,val))+'\\n')\n",
    "            else:\n",
    "                fh.write(line+'\\n')\n",
    "    fh.close()\n",
    "    \n",
    "    print ('Prepare %s data from project %s is completed !' %(dataFolders[genoDataType], project))\n",
    "\n",
    "    \n",
    "    \n",
    "########################################################\n",
    "for project in metaProjects.keys():\n",
    "    limmaInput(project)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "###################################################\n",
    "########## Step7. Parse Clinical data #############\n",
    "###################################################\n",
    "\n",
    "################# ALSO CAPABALE OF PARSING BIOSPECIMEN DATA, BUT NOT NECESSARY\n",
    "\n",
    "\n",
    "def parseClinical(project):\n",
    "    os.chdir(easyTCGA+project+'/'+dataFolders[genoDataType])\n",
    "    \n",
    "    fh = open(analysis+genoDataNames[genoDataType]+project+'.txt', 'wt')\n",
    "    \n",
    "    xmlFiles = [fl for fl in os.listdir(clean) if fl.endswith('xml')]\n",
    "    numSamples = len(xmlFiles)\n",
    "    \n",
    "    i=-1\n",
    "    clinicalData = OrderedDict()\n",
    "    for fl in xmlFiles:\n",
    "        i += 1\n",
    "        flName = fl.split('.')[0]\n",
    "        genoSamples.append(uuid[project][flName])\n",
    "        \n",
    "        doc = parse(clean+fl)\n",
    "        overlapTraits = []  ### check if the trait name has been used\n",
    "        for elem in doc.iter():\n",
    "            if elem.text == None:\n",
    "                elem.text = 'NA'\n",
    "            elif elem.text.startswith('\\n'):\n",
    "                continue\n",
    "            else:\n",
    "                trt = elem.tag.split('}')[1]\n",
    "                if overlapTraits.count(trt) != 0:\n",
    "                    trait = trt + '_' + str(overlapTraits.count(trt)+1)\n",
    "                else:\n",
    "                    trait = trt\n",
    "\n",
    "                overlapTraits.append(trt)\n",
    "                \n",
    "                if trait not in clinicalData.keys():\n",
    "                    clinicalData[trait] = ['NA']*numSamples\n",
    "                clinicalData[trait][i] = elem.text\n",
    "        \n",
    "    sampleIndex = [genoSamples.index(sam) for sam in samples[project]]\n",
    "    \n",
    "    fh.write('\\t'+'\\t'.join([genoSamples[i] for i in sampleIndex])+'\\n')\n",
    "    \n",
    "    for ke, val in clinicalData.items():\n",
    "        fh.write(ke+'\\t'+'\\t'.join([val[i] for i in sampleIndex])+'\\n')\n",
    "\n",
    "    fh.close()\n",
    "\n",
    "    \n",
    "    \n",
    "###############################################\n",
    "for project in metaProjects.keys():\n",
    "    parseClinical(project)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Name of the first metadata file:\n",
      "clinical.metadata.cart.2017-02-05T20-45-14.793906.1\n",
      "Name of the second metadata file:\n",
      "clinical.metadata.cart.2017-02-05T20-37-19.445296.2\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\n#### Method2, works, but the json file is a one-line file\\nmultiJSON = []\\nwith open(\\'RNAseq.metadata.cart.2017-02-03T22-17-14.782610.breast.lung.brain.bladder.prostate.ovary.json\\', \\'rt\\') as f:\\n    multiJSON.extend(json.load(f))\\n    \\nwith open(\\'RNAseq.metadata.cart.2017-02-03T22-23-54.182001.others.json\\', \\'rt\\') as f:\\n    multiJSON.extend(json.load(f))\\n\\nlen(multiJSON)\\n\\n\\nwith open(\"RNAseq222.TCGA.metadata.cart.2017-02-03.all.json\", \"wt\") as fh:\\n    json.dump(multiJSON, fh)\\n    \\n    \\n#### Method3, doesn\\'t work, need to manual check the connection line of the two files\\nos.system(\\'cat RNAseq.metadata.cart.2017-02-03T22-17-14.782610.breast.lung.brain.bladder.prostate.ovary.json           RNAseq.metadata.cart.2017-02-03T22-23-54.182001.others.json           > RNAseq333.TCGA.metadata.cart.2017-02-03.all.json\\')\\n'"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "################################################################\n",
    "############# Step0. Merge two metadata files ##################\n",
    "\n",
    "#### Method1, works, the same format as expected\n",
    "metaFile_1 = input('Name of the first metadata file:\\n')\n",
    "metaFile_2 = input('Name of the second metadata file:\\n')\n",
    "\n",
    "metaFile_new = re.sub('T\\d+\\S+', '.all', metaFile_1)\n",
    "\n",
    "fh = open(metaFile_new+'.json', 'wt')\n",
    "with open(metaFile_1+'.json', 'rt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        if line == '[':\n",
    "            lastLine = line\n",
    "            \n",
    "        elif line == ']':\n",
    "            fh.write(lastLine+',\\n')\n",
    "        else:\n",
    "            fh.write(lastLine+'\\n')\n",
    "            lastLine = line\n",
    "            \n",
    "with open(metaFile_2+'.json', 'rt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        if line == '[':\n",
    "            continue\n",
    "        fh.write(line+'\\n')\n",
    "\n",
    "fh.close()\n",
    "\n",
    "'''\n",
    "#### Method2, works, but the json file is a one-line file\n",
    "multiJSON = []\n",
    "with open('RNAseq.metadata.cart.2017-02-03T22-17-14.782610.breast.lung.brain.bladder.prostate.ovary.json', 'rt') as f:\n",
    "    multiJSON.extend(json.load(f))\n",
    "    \n",
    "with open('RNAseq.metadata.cart.2017-02-03T22-23-54.182001.others.json', 'rt') as f:\n",
    "    multiJSON.extend(json.load(f))\n",
    "\n",
    "len(multiJSON)\n",
    "\n",
    "\n",
    "with open(\"RNAseq222.TCGA.metadata.cart.2017-02-03.all.json\", \"wt\") as fh:\n",
    "    json.dump(multiJSON, fh)\n",
    "    \n",
    "    \n",
    "#### Method3, doesn't work, need to manual check the connection line of the two files\n",
    "os.system('cat RNAseq.metadata.cart.2017-02-03T22-17-14.782610.breast.lung.brain.bladder.prostate.ovary.json \\\n",
    "          RNAseq.metadata.cart.2017-02-03T22-23-54.182001.others.json \\\n",
    "          > RNAseq333.TCGA.metadata.cart.2017-02-03.all.json')\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#######################################\n",
    "########### ONLY USE FOR ONCE !!!\n",
    "\n",
    "###############################\n",
    "### Create all directories ####\n",
    "###############################\n",
    "import os\n",
    "os.chdir('/home/ruidong/Documents/Research/Cancer_Genomics/easyTCGA/')\n",
    "\n",
    "dataType = ['Gene_Expression_Quantification',\n",
    "            'Copy_Number_Segment',\n",
    "            'Masked_Copy_Number_Segment',\n",
    "            'Methylation_Beta_Value',\n",
    "            'Isoform_Expression_Quantification',\n",
    "            'miRNA_Expression_Quantification',\n",
    "            'Biospecimen_Supplement',\n",
    "            'Clinical_Supplement',\n",
    "            'Masked_Somatic_Mutation']\n",
    "\n",
    "\n",
    "with open('projects.2017-02-03T22-29-28.039066.txt', 'rt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        if line.startswith('ID'):\n",
    "            continue\n",
    "        \n",
    "        folders = ' '.join([lst[0]+'/'+datp+'/'+fd for datp in dataType for fd in ['Downloads','Clean','Analysis']])\n",
    "        os.system('mkdir -p '+folders)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Prepare miRNA_Expression_Quantification data from project TCGA-PRAD is completed !\n",
      "Prepare miRNA_Expression_Quantification data from project TCGA-BLCA is completed !\n",
      "Prepare miRNA_Expression_Quantification data from project TCGA-LGG is completed !\n"
     ]
    }
   ],
   "source": [
    "################################################################\n",
    "########## Step6. Prepare input for downstream analysis ######## FOR BACKUP\n",
    "################################################################\n",
    "\n",
    "### sample order is different from the orignial merged genotype file\n",
    "\n",
    "def limmaInput(project):\n",
    "    os.chdir(easyTCGA+project+'/'+dataFolders[genoDataType])\n",
    "    \n",
    "    fh = open(analysis+genoDataNames[genoDataType]+'_and_traits.'+project+'.txt', 'wt')\n",
    "\n",
    "    fh.write('\\t'+'\\t'.join(samples[project].keys())+'\\n')\n",
    "\n",
    "    for ke, val in traits[project].items():\n",
    "        fh.write(ke+'\\t'+'\\t'.join(map(str,val))+'\\n')\n",
    "\n",
    "\n",
    "    ### SHOULD CHANGE THE NAME FOR OTHER TYPES OF DATA\n",
    "    with open(analysis+genoDataNames[genoDataType]+'_only.'+project+'.txt', 'rt') as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            lst = line.split('\\t')\n",
    "            if lst[0] == '':\n",
    "                allSamples = lst[1:]\n",
    "                sampleIndex = [allSamples.index(sam) for sam in samples[project].keys()]\n",
    "            else:     #elif lst[0].startswith('ENSG|hsa|cg')\n",
    "                fh.write(lst[0]+'\\t'+'\\t'.join([lst[i+1] for i in sampleIndex])+'\\n')\n",
    "    \n",
    "    fh.close()\n",
    "    \n",
    "    print ('Prepare %s data from project %s is completed !' %(dataFolders[genoDataType], project))\n",
    "\n",
    "for project in metaProjects.keys():\n",
    "    limmaInput(project)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "### not in use\n",
    "def changeDirectory(project):\n",
    "    \n",
    "    if genoDataType == 1:\n",
    "        os.chdir(easyTCGA+project+'/'+RNAseq)\n",
    "    elif genoDataType == 2:\n",
    "        os.chdir(easyTCGA+project+'/'+miRNA)\n",
    "    elif genoDataType == 3:\n",
    "        os.chdir(easyTCGA+project+'/'+methylation)\n",
    "    elif genoDataType == 4:\n",
    "        os.chdir(easyTCGA+project+'/'+isoform)\n",
    "    elif genoDataType == 5:\n",
    "        os.chdir(easyTCGA+project+'/'+CNS)\n",
    "    elif genoDataType == 6:\n",
    "        os.chdir(easyTCGA+project+'/'+maskedCNS)"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
