{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "os.chdir('/home/ruidong/Documents/Research/Git/TCGA/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/ruidong/Documents/Research/Git/TCGA'"
      ]
     },
     "execution_count": 184,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "###########################################################################\n",
    "###                        Step 0. Setup Environment                    ###\n",
    "###########################################################################\n",
    "\n",
    "import re, os, scipy, json\n",
    "from os import system, path\n",
    "import zipfile\n",
    "import gzip\n",
    "import shutil\n",
    "import io\n",
    "import requests\n",
    "import json\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "\n",
    "from collections import OrderedDict, defaultdict\n",
    "from operator import itemgetter\n",
    "from pprint import pprint\n",
    "\n",
    "project = 'TCGA-CHOL'\n",
    "\n",
    "if not os.path.isdir(project):\n",
    "    os.mkdir(project)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "###########################################################################\n",
    "###                  Step 1. Download data by gdc-client                ###\n",
    "###########################################################################\n",
    "\n",
    "def gdcDownload(project, data_category, data_type, workflow_type, dest_dir, manifest=None):\n",
    "    \n",
    "    #manifest = input('Manifest file:')\n",
    "    #manifest = manifest.rstrip()\n",
    "    \n",
    "    #dest_dir = input('Destination directory:')\n",
    "    #dest_dir = dest_dir.rstrip()\n",
    "    \n",
    "    if not os.path.exists('gdc-client'):\n",
    "        os.system('wget https://gdc.cancer.gov/system/files/authenticated%20user/0/gdc-client_v1.3.0_Ubuntu14.04_x64.zip')\n",
    "        #os.system('unzip gdc-client_v1.3.0_Ubuntu14.04_x64.zip')\n",
    "    \n",
    "        with zipfile.ZipFile('gdc-client_v1.3.0_Ubuntu14.04_x64.zip','r') as zip_ref:\n",
    "            zip_ref.extractall('.')\n",
    "    \n",
    "    if not os.path.exists(dest_dir):\n",
    "        os.mkdir(dest_dir)\n",
    "    \n",
    "    \n",
    "    print ('Start Downloading ...')\n",
    "\n",
    "    \n",
    "    if manifest != None:\n",
    "        manifile = manifest\n",
    "        \n",
    "    else:\n",
    "        rm = True\n",
    "        manifest = getManifest(project, data_category, data_type, workflow_type)\n",
    "        manifile = '.'.join([data_category, data_type, workflow_type, project, 'txt'])\n",
    "        manifile = manifile.replace(' ', '_')\n",
    "        manifest.to_csv(manifile, header=True, index=False, sep='\\t')\n",
    "        \n",
    "    os.system('./gdc-client download -m '+ manifile)\n",
    "        \n",
    "        \n",
    "    fls = []\n",
    "    with open(manifile) as f:\n",
    "        for line in f:\n",
    "            line = line.rstrip()\n",
    "            lst = line.split('\\t')\n",
    "            \n",
    "            if lst[0] == 'id':\n",
    "                continue\n",
    "            else:\n",
    "                fls.append(lst[0])\n",
    "    \n",
    "    for fl in fls:\n",
    "        shutil.move(fl, dest_dir)\n",
    "        \n",
    "    if rm == True:\n",
    "        os.remove(manifile)\n",
    "    \n",
    "    print ('Download completed !')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start Downloading ...\n",
      "Download completed !\n"
     ]
    }
   ],
   "source": [
    "### Download data by providing manifest file\n",
    "\n",
    "data_type = 'miRNAs'\n",
    "dest_dir = os.path.join(project, data_type)\n",
    "\n",
    "manifest = 'gdc_manifest.Isoform.TCGA-CHOL.txt'\n",
    "\n",
    "gdcDownload(manifest, dest_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start Downloading ...\n",
      "Download completed !\n"
     ]
    }
   ],
   "source": [
    "### Download data by specifying project, data category, etc.\n",
    "\n",
    "data_type = 'RNAseq'\n",
    "dest_dir = os.path.join(project, data_type)\n",
    "\n",
    "gdcDownload(project = 'TCGA-CHOL', \n",
    "            data_category = 'Transcriptome Profiling', \n",
    "            data_type = 'Gene Expression Quantification', \n",
    "            workflow_type = 'HTSeq - Counts',\n",
    "            dest_dir = dest_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Start Downloading ...\n",
      "Download completed !\n"
     ]
    }
   ],
   "source": [
    "data_type = 'preMIR'\n",
    "dest_dir = os.path.join(project, data_type)\n",
    "\n",
    "gdcDownload(project = 'TCGA-CHOL', \n",
    "            data_category = 'Transcriptome Profiling', \n",
    "            data_type = 'miRNA Expression Quantification', \n",
    "            workflow_type = 'BCGSC miRNA Profiling',\n",
    "            dest_dir = dest_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "data_type = 'Methylation'\n",
    "dest_dir = os.path.join(project, data_type)\n",
    "\n",
    "gdcDownload(project = 'TCGA-CHOL', \n",
    "            data_category = 'DNA Methylation', \n",
    "            data_type = 'Methylation Beta Value', \n",
    "            workflow_type = 'Liftover',\n",
    "            dest_dir = dest_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "###########################################################################\n",
    "###                 Step 2. Parse metadata via GDC API                  ###\n",
    "###########################################################################\n",
    "\n",
    "\n",
    "def gdcParseMeta(project, data_category, data_type, workflow_type):\n",
    "    \n",
    "    files_endpt = \"https://api.gdc.cancer.gov/files\"\n",
    "    \n",
    "    fields = [\n",
    "        \"file_name\",\n",
    "        #\"file_id\",\n",
    "        \"cases.samples.submitter_id\",\n",
    "        \"associated_entities.entity_submitter_id\",\n",
    "        \"cases.samples.sample_type\",\n",
    "        \"cases.disease_type\",\n",
    "        \"cases.demographic.gender\",\n",
    "        \"cases.diagnoses.tumor_stage\",\n",
    "        \"cases.diagnoses.tumor_grade\",\n",
    "        \"cases.diagnoses.age_at_diagnosis\",\n",
    "        \"cases.diagnoses.days_to_death\",\n",
    "        \"cases.diagnoses.days_to_last_follow_up\",\n",
    "        \"cases.diagnoses.vital_status\",\n",
    "        \"cases.project.project_id\"\n",
    "    ]\n",
    "\n",
    "    fields = \",\".join(fields)\n",
    "\n",
    "    # This set of filters is nested under an 'and' operator.\n",
    "    filters = {\n",
    "        \"op\": \"and\",\n",
    "        \"content\":[\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"cases.project.project_id\",\n",
    "                \"value\": [project]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"files.data_category\",\n",
    "                \"value\": [data_category]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"files.data_type\",\n",
    "                \"value\": [data_type]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"files.analysis.workflow_type\",\n",
    "                \"value\": [workflow_type]\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "    \n",
    "    # A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "    params = {\n",
    "        \"filters\": filters,\n",
    "        \"fields\": fields,\n",
    "        \"format\": \"TSV\",\n",
    "        \"size\": \"10000\"\n",
    "        }\n",
    "    \n",
    "    # The parameters are passed to 'json' rather than 'params' in this case\n",
    "    response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "    \n",
    "    df = pd.read_csv(io.StringIO(response.content.decode(\"utf-8\")), sep='\\t')\n",
    "    \n",
    "    df.columns = ['days_to_death', 'age_at_diagnosis', 'tumor_grade',\n",
    "                 'sample_type', 'tumor_stage', 'entity_submitter_id',\n",
    "                 'submitter_id', 'project_id', 'vital_status','disease_type', \n",
    "                 'file_name', 'days_to_last_follow_up','gender', 'file_id']\n",
    "    \n",
    "    df = pd.DataFrame(df, columns = ['file_id', 'file_name', 'submitter_id', 'entity_submitter_id',\n",
    "                                     'sample_type', 'age_at_diagnosis', 'gender', 'tumor_stage',\n",
    "                                     'tumor_grade', 'days_to_death', 'days_to_last_follow_up',\n",
    "                                     'vital_status', 'disease_type', 'project_id'])\n",
    "    \n",
    "    df = df.rename_axis(df['submitter_id'], axis=\"rows\")\n",
    "    #print(response.content.decode(\"utf-8\"))\n",
    "    return (df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 219,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "meta = gdcParseMeta(project = 'TCGA-CHOL', \n",
    "                    data_category = 'Transcriptome Profiling', \n",
    "                    data_type = 'Isoform Expression Quantification', \n",
    "                    workflow_type = 'BCGSC miRNA Profiling')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id</th>\n",
       "      <th>file_name</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>entity_submitter_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>tumor_stage</th>\n",
       "      <th>tumor_grade</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_follow_up</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>project_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA36-01A</th>\n",
       "      <td>40a40b43-142a-4041-beac-e4c55dfab09d</td>\n",
       "      <td>91c07e4a-3521-4f9c-a1ad-ba8c677a7c98.mirbase21...</td>\n",
       "      <td>TCGA-W5-AA36-01A</td>\n",
       "      <td>TCGA-W5-AA36-01A-11R-A41D-13</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>18882</td>\n",
       "      <td>female</td>\n",
       "      <td>stage iv</td>\n",
       "      <td>not reported</td>\n",
       "      <td>1402.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-YR-A95A-01A</th>\n",
       "      <td>2370512a-bdd1-4554-b5e2-421e4a44c5a1</td>\n",
       "      <td>07a10aa5-fb68-4d78-b02c-f1456e297f4f.mirbase21...</td>\n",
       "      <td>TCGA-YR-A95A-01A</td>\n",
       "      <td>TCGA-YR-A95A-01A-12R-A41D-13</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>19292</td>\n",
       "      <td>male</td>\n",
       "      <td>stage iv</td>\n",
       "      <td>not reported</td>\n",
       "      <td>26.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3X-AAV9-01A</th>\n",
       "      <td>85f0797e-1d52-461e-a9f0-519c26908a04</td>\n",
       "      <td>32100d21-e578-43a9-9e7e-0a59069fd6bf.mirbase21...</td>\n",
       "      <td>TCGA-3X-AAV9-01A</td>\n",
       "      <td>TCGA-3X-AAV9-01A-72R-A41D-13</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>26349</td>\n",
       "      <td>male</td>\n",
       "      <td>stage i</td>\n",
       "      <td>not reported</td>\n",
       "      <td>339.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA2G-01A</th>\n",
       "      <td>30cf8b89-708a-40df-9d4e-72eab593ac13</td>\n",
       "      <td>44ad6046-e17a-4655-bb58-3c778930e938.mirbase21...</td>\n",
       "      <td>TCGA-W5-AA2G-01A</td>\n",
       "      <td>TCGA-W5-AA2G-01A-11R-A41D-13</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>22933</td>\n",
       "      <td>female</td>\n",
       "      <td>stage i</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1976.0</td>\n",
       "      <td>alive</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA2Q-01A</th>\n",
       "      <td>aeecfc69-fdf7-450c-b374-0f7a21c74da5</td>\n",
       "      <td>caad916f-d439-4849-a3a5-e29dcd54cd43.mirbase21...</td>\n",
       "      <td>TCGA-W5-AA2Q-01A</td>\n",
       "      <td>TCGA-W5-AA2Q-01A-11R-A41D-13</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>25069</td>\n",
       "      <td>male</td>\n",
       "      <td>stage ii</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50.0</td>\n",
       "      <td>alive</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               file_id  \\\n",
       "TCGA-W5-AA36-01A  40a40b43-142a-4041-beac-e4c55dfab09d   \n",
       "TCGA-YR-A95A-01A  2370512a-bdd1-4554-b5e2-421e4a44c5a1   \n",
       "TCGA-3X-AAV9-01A  85f0797e-1d52-461e-a9f0-519c26908a04   \n",
       "TCGA-W5-AA2G-01A  30cf8b89-708a-40df-9d4e-72eab593ac13   \n",
       "TCGA-W5-AA2Q-01A  aeecfc69-fdf7-450c-b374-0f7a21c74da5   \n",
       "\n",
       "                                                          file_name  \\\n",
       "TCGA-W5-AA36-01A  91c07e4a-3521-4f9c-a1ad-ba8c677a7c98.mirbase21...   \n",
       "TCGA-YR-A95A-01A  07a10aa5-fb68-4d78-b02c-f1456e297f4f.mirbase21...   \n",
       "TCGA-3X-AAV9-01A  32100d21-e578-43a9-9e7e-0a59069fd6bf.mirbase21...   \n",
       "TCGA-W5-AA2G-01A  44ad6046-e17a-4655-bb58-3c778930e938.mirbase21...   \n",
       "TCGA-W5-AA2Q-01A  caad916f-d439-4849-a3a5-e29dcd54cd43.mirbase21...   \n",
       "\n",
       "                      submitter_id           entity_submitter_id  \\\n",
       "TCGA-W5-AA36-01A  TCGA-W5-AA36-01A  TCGA-W5-AA36-01A-11R-A41D-13   \n",
       "TCGA-YR-A95A-01A  TCGA-YR-A95A-01A  TCGA-YR-A95A-01A-12R-A41D-13   \n",
       "TCGA-3X-AAV9-01A  TCGA-3X-AAV9-01A  TCGA-3X-AAV9-01A-72R-A41D-13   \n",
       "TCGA-W5-AA2G-01A  TCGA-W5-AA2G-01A  TCGA-W5-AA2G-01A-11R-A41D-13   \n",
       "TCGA-W5-AA2Q-01A  TCGA-W5-AA2Q-01A  TCGA-W5-AA2Q-01A-11R-A41D-13   \n",
       "\n",
       "                    sample_type  age_at_diagnosis  gender tumor_stage  \\\n",
       "TCGA-W5-AA36-01A  Primary Tumor             18882  female    stage iv   \n",
       "TCGA-YR-A95A-01A  Primary Tumor             19292    male    stage iv   \n",
       "TCGA-3X-AAV9-01A  Primary Tumor             26349    male     stage i   \n",
       "TCGA-W5-AA2G-01A  Primary Tumor             22933  female     stage i   \n",
       "TCGA-W5-AA2Q-01A  Primary Tumor             25069    male    stage ii   \n",
       "\n",
       "                   tumor_grade  days_to_death  days_to_last_follow_up  \\\n",
       "TCGA-W5-AA36-01A  not reported         1402.0                     NaN   \n",
       "TCGA-YR-A95A-01A  not reported           26.0                     NaN   \n",
       "TCGA-3X-AAV9-01A  not reported          339.0                     NaN   \n",
       "TCGA-W5-AA2G-01A  not reported            NaN                  1976.0   \n",
       "TCGA-W5-AA2Q-01A  not reported            NaN                    50.0   \n",
       "\n",
       "                 vital_status        disease_type project_id  \n",
       "TCGA-W5-AA36-01A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-YR-A95A-01A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-3X-AAV9-01A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA2G-01A        alive  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA2Q-01A        alive  Cholangiocarcinoma  TCGA-CHOL  "
      ]
     },
     "execution_count": 211,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "###########################################################################\n",
    "###               Step 3. Merge data for downstream analysis            ###\n",
    "###########################################################################\n",
    "\n",
    "def gdcMerge(path, meta, data_type): \n",
    "    \n",
    "    fls = meta['file_id'] + '/' + meta['file_name']\n",
    "    sams = fls.to_dict()\n",
    "    \n",
    "    expr = {}\n",
    "    \n",
    "    ### mature miRNAs\n",
    "    if data_type == 'miRNAs':\n",
    "        \n",
    "        mir21 = OrderedDict()\n",
    "    \n",
    "        with open('mature.release21.id.txt') as f:\n",
    "            for line in f:\n",
    "                line = line.rstrip()\n",
    "                lst = line.split('\\t')\n",
    "                mir21[lst[0]] = lst[1]\n",
    "\n",
    "        for sam,fl in sams.items():\n",
    "            expr[sam] = OrderedDict()\n",
    "    \n",
    "            for mir in mir21.values():\n",
    "                expr[sam][mir] = 0\n",
    "    \n",
    "            with open(path+'/'+fl) as f:\n",
    "                for line in f:\n",
    "                    line = line.rstrip()\n",
    "                    lst = line.split('\\t')\n",
    "            \n",
    "                    if lst[-1].startswith('mature'):\n",
    "                        mir = lst[-1].split(',')[-1]\n",
    "                        expr[sam][mir21[mir]] += int(lst[2])\n",
    "    \n",
    "    ### RNAseq\n",
    "    if data_type == 'RNAseq':\n",
    "        for sam,fl in sams.items():\n",
    "            expr[sam] = OrderedDict()\n",
    "            \n",
    "            with gzip.open(path+'/'+fl, 'rt') as f:\n",
    "                for line in f:\n",
    "                    line = line.rstrip()\n",
    "                    lst = line.split('\\t')\n",
    "            \n",
    "                    if not lst[0].startswith('ENSG'):\n",
    "                        continue\n",
    "                        \n",
    "                    gene = lst[0].split('.')[0]\n",
    "                        \n",
    "                    expr[sam][gene] = lst[-1]\n",
    "    \n",
    "    \n",
    "    ### Methylation and precursor miRNAs\n",
    "    if data_type in ['Methylation', 'preMIR']:\n",
    "        for sam,fl in sams.items():\n",
    "            expr[sam] = OrderedDict()\n",
    "            \n",
    "            with open(path+'/'+fl) as f:\n",
    "                for line in f:\n",
    "                    line = line.rstrip()\n",
    "                    lst = line.split('\\t')\n",
    "            \n",
    "                    if lst[0].startswith('Composite|miRNA_ID'):\n",
    "                        continue\n",
    "                        \n",
    "                    expr[sam][lst[0]] = lst[1]        \n",
    "    \n",
    "                    \n",
    "    mergeDa = pd.DataFrame.from_dict(expr)\n",
    "        \n",
    "    return (mergeDa)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 220,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "mirExpr = gdcMerge(path='TCGA-CHOL/miRNAs', meta=meta, data_type='miRNAs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 221,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-3X-AAV9-01A</th>\n",
       "      <th>TCGA-3X-AAVA-01A</th>\n",
       "      <th>TCGA-3X-AAVB-01A</th>\n",
       "      <th>TCGA-3X-AAVC-01A</th>\n",
       "      <th>TCGA-3X-AAVE-01A</th>\n",
       "      <th>TCGA-4G-AAZO-01A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-2-3p</th>\n",
       "      <td>29</td>\n",
       "      <td>25</td>\n",
       "      <td>46</td>\n",
       "      <td>1</td>\n",
       "      <td>38</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-3p</th>\n",
       "      <td>188</td>\n",
       "      <td>155</td>\n",
       "      <td>282</td>\n",
       "      <td>120</td>\n",
       "      <td>141</td>\n",
       "      <td>210</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-5p</th>\n",
       "      <td>153811</td>\n",
       "      <td>119809</td>\n",
       "      <td>190562</td>\n",
       "      <td>79303</td>\n",
       "      <td>110684</td>\n",
       "      <td>107689</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7b-3p</th>\n",
       "      <td>45</td>\n",
       "      <td>58</td>\n",
       "      <td>129</td>\n",
       "      <td>60</td>\n",
       "      <td>115</td>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7b-5p</th>\n",
       "      <td>58883</td>\n",
       "      <td>34952</td>\n",
       "      <td>123675</td>\n",
       "      <td>51800</td>\n",
       "      <td>94827</td>\n",
       "      <td>104602</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7c-3p</th>\n",
       "      <td>40</td>\n",
       "      <td>34</td>\n",
       "      <td>48</td>\n",
       "      <td>9</td>\n",
       "      <td>34</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 TCGA-3X-AAV9-01A  TCGA-3X-AAVA-01A  TCGA-3X-AAVB-01A  \\\n",
       "hsa-let-7a-2-3p                29                25                46   \n",
       "hsa-let-7a-3p                 188               155               282   \n",
       "hsa-let-7a-5p              153811            119809            190562   \n",
       "hsa-let-7b-3p                  45                58               129   \n",
       "hsa-let-7b-5p               58883             34952            123675   \n",
       "hsa-let-7c-3p                  40                34                48   \n",
       "\n",
       "                 TCGA-3X-AAVC-01A  TCGA-3X-AAVE-01A  TCGA-4G-AAZO-01A  \n",
       "hsa-let-7a-2-3p                 1                38                30  \n",
       "hsa-let-7a-3p                 120               141               210  \n",
       "hsa-let-7a-5p               79303            110684            107689  \n",
       "hsa-let-7b-3p                  60               115                95  \n",
       "hsa-let-7b-5p               51800             94827            104602  \n",
       "hsa-let-7c-3p                   9                34                31  "
      ]
     },
     "execution_count": 221,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mirExpr.iloc[:6, :6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 222,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "meta = gdcParseMeta(project = 'TCGA-CHOL', \n",
    "                    data_category = 'Transcriptome Profiling', \n",
    "                    data_type = 'Gene Expression Quantification', \n",
    "                    workflow_type = 'HTSeq - Counts')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id</th>\n",
       "      <th>file_name</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>entity_submitter_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>tumor_stage</th>\n",
       "      <th>tumor_grade</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_follow_up</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>project_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA2R-01A</th>\n",
       "      <td>afecdda2-735c-4304-a087-ef917ad9cd5a</td>\n",
       "      <td>fc06a930-e3a1-4775-a126-610a531c655b.htseq.cou...</td>\n",
       "      <td>TCGA-W5-AA2R-01A</td>\n",
       "      <td>TCGA-W5-AA2R-01A-11R-A41I-07</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>28367</td>\n",
       "      <td>female</td>\n",
       "      <td>stage i</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1542.0</td>\n",
       "      <td>alive</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA2U-01A</th>\n",
       "      <td>4963b9af-7d7b-42cd-b057-2f894639e59f</td>\n",
       "      <td>46fd54e8-5ab7-43f1-bb88-01a163ff121f.htseq.cou...</td>\n",
       "      <td>TCGA-W5-AA2U-01A</td>\n",
       "      <td>TCGA-W5-AA2U-01A-11R-A41I-07</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>28552</td>\n",
       "      <td>female</td>\n",
       "      <td>stage i</td>\n",
       "      <td>not reported</td>\n",
       "      <td>627.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA34-11A</th>\n",
       "      <td>d9249d25-d852-4008-a71e-dd9be6cb68e9</td>\n",
       "      <td>fbdb830c-2f70-45f5-a2d6-2d8317301af3.htseq.cou...</td>\n",
       "      <td>TCGA-W5-AA34-11A</td>\n",
       "      <td>TCGA-W5-AA34-11A-11R-A41I-07</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>27535</td>\n",
       "      <td>female</td>\n",
       "      <td>stage i</td>\n",
       "      <td>not reported</td>\n",
       "      <td>555.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-ZH-A8Y6-01A</th>\n",
       "      <td>be2d52c8-eb14-4b45-a65b-c1bdd6f5c4c2</td>\n",
       "      <td>4696ce44-29bf-41ea-b866-ddb17c376e94.htseq.cou...</td>\n",
       "      <td>TCGA-ZH-A8Y6-01A</td>\n",
       "      <td>TCGA-ZH-A8Y6-01A-11R-A41I-07</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>14976</td>\n",
       "      <td>female</td>\n",
       "      <td>stage i</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>519.0</td>\n",
       "      <td>alive</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3X-AAVA-01A</th>\n",
       "      <td>42b8d463-6209-4ea0-bb01-8023a1302fa0</td>\n",
       "      <td>b6a2c03a-c8ad-41e9-8a19-8f5ac53cae9f.htseq.cou...</td>\n",
       "      <td>TCGA-3X-AAVA-01A</td>\n",
       "      <td>TCGA-3X-AAVA-01A-11R-A41I-07</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>18303</td>\n",
       "      <td>female</td>\n",
       "      <td>stage ii</td>\n",
       "      <td>not reported</td>\n",
       "      <td>445.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               file_id  \\\n",
       "TCGA-W5-AA2R-01A  afecdda2-735c-4304-a087-ef917ad9cd5a   \n",
       "TCGA-W5-AA2U-01A  4963b9af-7d7b-42cd-b057-2f894639e59f   \n",
       "TCGA-W5-AA34-11A  d9249d25-d852-4008-a71e-dd9be6cb68e9   \n",
       "TCGA-ZH-A8Y6-01A  be2d52c8-eb14-4b45-a65b-c1bdd6f5c4c2   \n",
       "TCGA-3X-AAVA-01A  42b8d463-6209-4ea0-bb01-8023a1302fa0   \n",
       "\n",
       "                                                          file_name  \\\n",
       "TCGA-W5-AA2R-01A  fc06a930-e3a1-4775-a126-610a531c655b.htseq.cou...   \n",
       "TCGA-W5-AA2U-01A  46fd54e8-5ab7-43f1-bb88-01a163ff121f.htseq.cou...   \n",
       "TCGA-W5-AA34-11A  fbdb830c-2f70-45f5-a2d6-2d8317301af3.htseq.cou...   \n",
       "TCGA-ZH-A8Y6-01A  4696ce44-29bf-41ea-b866-ddb17c376e94.htseq.cou...   \n",
       "TCGA-3X-AAVA-01A  b6a2c03a-c8ad-41e9-8a19-8f5ac53cae9f.htseq.cou...   \n",
       "\n",
       "                      submitter_id           entity_submitter_id  \\\n",
       "TCGA-W5-AA2R-01A  TCGA-W5-AA2R-01A  TCGA-W5-AA2R-01A-11R-A41I-07   \n",
       "TCGA-W5-AA2U-01A  TCGA-W5-AA2U-01A  TCGA-W5-AA2U-01A-11R-A41I-07   \n",
       "TCGA-W5-AA34-11A  TCGA-W5-AA34-11A  TCGA-W5-AA34-11A-11R-A41I-07   \n",
       "TCGA-ZH-A8Y6-01A  TCGA-ZH-A8Y6-01A  TCGA-ZH-A8Y6-01A-11R-A41I-07   \n",
       "TCGA-3X-AAVA-01A  TCGA-3X-AAVA-01A  TCGA-3X-AAVA-01A-11R-A41I-07   \n",
       "\n",
       "                          sample_type  age_at_diagnosis  gender tumor_stage  \\\n",
       "TCGA-W5-AA2R-01A        Primary Tumor             28367  female     stage i   \n",
       "TCGA-W5-AA2U-01A        Primary Tumor             28552  female     stage i   \n",
       "TCGA-W5-AA34-11A  Solid Tissue Normal             27535  female     stage i   \n",
       "TCGA-ZH-A8Y6-01A        Primary Tumor             14976  female     stage i   \n",
       "TCGA-3X-AAVA-01A        Primary Tumor             18303  female    stage ii   \n",
       "\n",
       "                   tumor_grade  days_to_death  days_to_last_follow_up  \\\n",
       "TCGA-W5-AA2R-01A  not reported            NaN                  1542.0   \n",
       "TCGA-W5-AA2U-01A  not reported          627.0                     NaN   \n",
       "TCGA-W5-AA34-11A  not reported          555.0                   168.0   \n",
       "TCGA-ZH-A8Y6-01A  not reported            NaN                   519.0   \n",
       "TCGA-3X-AAVA-01A  not reported          445.0                     NaN   \n",
       "\n",
       "                 vital_status        disease_type project_id  \n",
       "TCGA-W5-AA2R-01A        alive  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA2U-01A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA34-11A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-ZH-A8Y6-01A        alive  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-3X-AAVA-01A         dead  Cholangiocarcinoma  TCGA-CHOL  "
      ]
     },
     "execution_count": 146,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 223,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "rnaExpr = gdcMerge(path='TCGA-CHOL/RNAseq', meta=meta, data_type='RNAseq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-3X-AAV9-01A</th>\n",
       "      <th>TCGA-3X-AAVA-01A</th>\n",
       "      <th>TCGA-3X-AAVB-01A</th>\n",
       "      <th>TCGA-3X-AAVC-01A</th>\n",
       "      <th>TCGA-3X-AAVE-01A</th>\n",
       "      <th>TCGA-4G-AAZO-01A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>3363</td>\n",
       "      <td>4262</td>\n",
       "      <td>2001</td>\n",
       "      <td>3960</td>\n",
       "      <td>4798</td>\n",
       "      <td>1631</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>1804</td>\n",
       "      <td>1254</td>\n",
       "      <td>1365</td>\n",
       "      <td>1250</td>\n",
       "      <td>1492</td>\n",
       "      <td>757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>601</td>\n",
       "      <td>699</td>\n",
       "      <td>584</td>\n",
       "      <td>1127</td>\n",
       "      <td>1110</td>\n",
       "      <td>1535</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>242</td>\n",
       "      <td>239</td>\n",
       "      <td>238</td>\n",
       "      <td>217</td>\n",
       "      <td>368</td>\n",
       "      <td>162</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>670</td>\n",
       "      <td>334</td>\n",
       "      <td>484</td>\n",
       "      <td>421</td>\n",
       "      <td>696</td>\n",
       "      <td>137</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                TCGA-3X-AAV9-01A TCGA-3X-AAVA-01A TCGA-3X-AAVB-01A  \\\n",
       "ENSG00000000003             3363             4262             2001   \n",
       "ENSG00000000005                1                1                2   \n",
       "ENSG00000000419             1804             1254             1365   \n",
       "ENSG00000000457              601              699              584   \n",
       "ENSG00000000460              242              239              238   \n",
       "ENSG00000000938              670              334              484   \n",
       "\n",
       "                TCGA-3X-AAVC-01A TCGA-3X-AAVE-01A TCGA-4G-AAZO-01A  \n",
       "ENSG00000000003             3960             4798             1631  \n",
       "ENSG00000000005                1                0                0  \n",
       "ENSG00000000419             1250             1492              757  \n",
       "ENSG00000000457             1127             1110             1535  \n",
       "ENSG00000000460              217              368              162  \n",
       "ENSG00000000938              421              696              137  "
      ]
     },
     "execution_count": 224,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rnaExpr.iloc[:6, :6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id</th>\n",
       "      <th>file_name</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>entity_submitter_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>tumor_stage</th>\n",
       "      <th>tumor_grade</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_follow_up</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>project_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-ZU-A8S4-11A</th>\n",
       "      <td>4d7b448e-496d-474d-856e-ee578a6559a5</td>\n",
       "      <td>78b783b1-c2eb-4b94-9765-b4b8bf3776e0.mirbase21...</td>\n",
       "      <td>TCGA-ZU-A8S4-11A</td>\n",
       "      <td>TCGA-ZU-A8S4-11A-11R-A41D-13</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>19264</td>\n",
       "      <td>male</td>\n",
       "      <td>stage i</td>\n",
       "      <td>not reported</td>\n",
       "      <td>98.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA34-11A</th>\n",
       "      <td>6c78a05b-c1e8-4c20-adfc-262a95b4ea49</td>\n",
       "      <td>f2809fd4-3e86-4ee2-ae5a-6a56256b95e7.mirbase21...</td>\n",
       "      <td>TCGA-W5-AA34-11A</td>\n",
       "      <td>TCGA-W5-AA34-11A-11R-A41D-13</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>27535</td>\n",
       "      <td>female</td>\n",
       "      <td>stage i</td>\n",
       "      <td>not reported</td>\n",
       "      <td>555.0</td>\n",
       "      <td>168.0</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA2T-01A</th>\n",
       "      <td>15bfe93a-9236-4ff0-9145-78a7bbf22cb6</td>\n",
       "      <td>2ab22567-023f-4e2b-80a5-00c7d58f4716.mirbase21...</td>\n",
       "      <td>TCGA-W5-AA2T-01A</td>\n",
       "      <td>TCGA-W5-AA2T-01A-12R-A41D-13</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>23560</td>\n",
       "      <td>female</td>\n",
       "      <td>stage ii</td>\n",
       "      <td>not reported</td>\n",
       "      <td>1220.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA2Q-01A</th>\n",
       "      <td>74d60a5e-cd0b-4291-90df-35a42f476d78</td>\n",
       "      <td>caad916f-d439-4849-a3a5-e29dcd54cd43.mirbase21...</td>\n",
       "      <td>TCGA-W5-AA2Q-01A</td>\n",
       "      <td>TCGA-W5-AA2Q-01A-11R-A41D-13</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>25069</td>\n",
       "      <td>male</td>\n",
       "      <td>stage ii</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "      <td>50.0</td>\n",
       "      <td>alive</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA2O-01A</th>\n",
       "      <td>455ee4cb-72f9-41f8-908e-ba15b92fafd9</td>\n",
       "      <td>88ef0233-d4f6-43ee-aa5f-2bb5089072d5.mirbase21...</td>\n",
       "      <td>TCGA-W5-AA2O-01A</td>\n",
       "      <td>TCGA-W5-AA2O-01A-11R-A41D-13</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>21118</td>\n",
       "      <td>male</td>\n",
       "      <td>stage i</td>\n",
       "      <td>not reported</td>\n",
       "      <td>640.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               file_id  \\\n",
       "TCGA-ZU-A8S4-11A  4d7b448e-496d-474d-856e-ee578a6559a5   \n",
       "TCGA-W5-AA34-11A  6c78a05b-c1e8-4c20-adfc-262a95b4ea49   \n",
       "TCGA-W5-AA2T-01A  15bfe93a-9236-4ff0-9145-78a7bbf22cb6   \n",
       "TCGA-W5-AA2Q-01A  74d60a5e-cd0b-4291-90df-35a42f476d78   \n",
       "TCGA-W5-AA2O-01A  455ee4cb-72f9-41f8-908e-ba15b92fafd9   \n",
       "\n",
       "                                                          file_name  \\\n",
       "TCGA-ZU-A8S4-11A  78b783b1-c2eb-4b94-9765-b4b8bf3776e0.mirbase21...   \n",
       "TCGA-W5-AA34-11A  f2809fd4-3e86-4ee2-ae5a-6a56256b95e7.mirbase21...   \n",
       "TCGA-W5-AA2T-01A  2ab22567-023f-4e2b-80a5-00c7d58f4716.mirbase21...   \n",
       "TCGA-W5-AA2Q-01A  caad916f-d439-4849-a3a5-e29dcd54cd43.mirbase21...   \n",
       "TCGA-W5-AA2O-01A  88ef0233-d4f6-43ee-aa5f-2bb5089072d5.mirbase21...   \n",
       "\n",
       "                      submitter_id           entity_submitter_id  \\\n",
       "TCGA-ZU-A8S4-11A  TCGA-ZU-A8S4-11A  TCGA-ZU-A8S4-11A-11R-A41D-13   \n",
       "TCGA-W5-AA34-11A  TCGA-W5-AA34-11A  TCGA-W5-AA34-11A-11R-A41D-13   \n",
       "TCGA-W5-AA2T-01A  TCGA-W5-AA2T-01A  TCGA-W5-AA2T-01A-12R-A41D-13   \n",
       "TCGA-W5-AA2Q-01A  TCGA-W5-AA2Q-01A  TCGA-W5-AA2Q-01A-11R-A41D-13   \n",
       "TCGA-W5-AA2O-01A  TCGA-W5-AA2O-01A  TCGA-W5-AA2O-01A-11R-A41D-13   \n",
       "\n",
       "                          sample_type  age_at_diagnosis  gender tumor_stage  \\\n",
       "TCGA-ZU-A8S4-11A  Solid Tissue Normal             19264    male     stage i   \n",
       "TCGA-W5-AA34-11A  Solid Tissue Normal             27535  female     stage i   \n",
       "TCGA-W5-AA2T-01A        Primary Tumor             23560  female    stage ii   \n",
       "TCGA-W5-AA2Q-01A        Primary Tumor             25069    male    stage ii   \n",
       "TCGA-W5-AA2O-01A        Primary Tumor             21118    male     stage i   \n",
       "\n",
       "                   tumor_grade  days_to_death  days_to_last_follow_up  \\\n",
       "TCGA-ZU-A8S4-11A  not reported           98.0                     NaN   \n",
       "TCGA-W5-AA34-11A  not reported          555.0                   168.0   \n",
       "TCGA-W5-AA2T-01A  not reported         1220.0                     NaN   \n",
       "TCGA-W5-AA2Q-01A  not reported            NaN                    50.0   \n",
       "TCGA-W5-AA2O-01A  not reported          640.0                     NaN   \n",
       "\n",
       "                 vital_status        disease_type project_id  \n",
       "TCGA-ZU-A8S4-11A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA34-11A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA2T-01A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA2Q-01A        alive  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA2O-01A         dead  Cholangiocarcinoma  TCGA-CHOL  "
      ]
     },
     "execution_count": 225,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta = gdcParseMeta(project = 'TCGA-CHOL', \n",
    "                    data_category = 'Transcriptome Profiling', \n",
    "                    data_type = 'miRNA Expression Quantification', \n",
    "                    workflow_type = 'BCGSC miRNA Profiling')\n",
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "preMIR = gdcMerge(path='TCGA-CHOL/preMIR', meta=meta, data_type='preMIR')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-3X-AAV9-01A</th>\n",
       "      <th>TCGA-3X-AAVA-01A</th>\n",
       "      <th>TCGA-3X-AAVB-01A</th>\n",
       "      <th>TCGA-3X-AAVC-01A</th>\n",
       "      <th>TCGA-3X-AAVE-01A</th>\n",
       "      <th>TCGA-4G-AAZO-01A</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-1</th>\n",
       "      <td>51254</td>\n",
       "      <td>39980</td>\n",
       "      <td>63775</td>\n",
       "      <td>26495</td>\n",
       "      <td>36932</td>\n",
       "      <td>36191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-2</th>\n",
       "      <td>51237</td>\n",
       "      <td>39889</td>\n",
       "      <td>63280</td>\n",
       "      <td>26414</td>\n",
       "      <td>36851</td>\n",
       "      <td>35811</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-3</th>\n",
       "      <td>51547</td>\n",
       "      <td>40122</td>\n",
       "      <td>63843</td>\n",
       "      <td>26516</td>\n",
       "      <td>37080</td>\n",
       "      <td>35932</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7b</th>\n",
       "      <td>58928</td>\n",
       "      <td>35010</td>\n",
       "      <td>123804</td>\n",
       "      <td>51860</td>\n",
       "      <td>94942</td>\n",
       "      <td>104700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7c</th>\n",
       "      <td>13027</td>\n",
       "      <td>3246</td>\n",
       "      <td>9751</td>\n",
       "      <td>760</td>\n",
       "      <td>4719</td>\n",
       "      <td>2551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7d</th>\n",
       "      <td>1229</td>\n",
       "      <td>2536</td>\n",
       "      <td>2294</td>\n",
       "      <td>2718</td>\n",
       "      <td>2356</td>\n",
       "      <td>1317</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             TCGA-3X-AAV9-01A TCGA-3X-AAVA-01A TCGA-3X-AAVB-01A  \\\n",
       "hsa-let-7a-1            51254            39980            63775   \n",
       "hsa-let-7a-2            51237            39889            63280   \n",
       "hsa-let-7a-3            51547            40122            63843   \n",
       "hsa-let-7b              58928            35010           123804   \n",
       "hsa-let-7c              13027             3246             9751   \n",
       "hsa-let-7d               1229             2536             2294   \n",
       "\n",
       "             TCGA-3X-AAVC-01A TCGA-3X-AAVE-01A TCGA-4G-AAZO-01A  \n",
       "hsa-let-7a-1            26495            36932            36191  \n",
       "hsa-let-7a-2            26414            36851            35811  \n",
       "hsa-let-7a-3            26516            37080            35932  \n",
       "hsa-let-7b              51860            94942           104700  \n",
       "hsa-let-7c                760             4719             2551  \n",
       "hsa-let-7d               2718             2356             1317  "
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "preMIR.iloc[:6, :6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "meta = gdcParseMeta(project = 'TCGA-CHOL', \n",
    "                    data_category = 'DNA Methylation', \n",
    "                    data_type = 'Methylation Beta Value', \n",
    "                    workflow_type = 'Liftover')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>file_id</th>\n",
       "      <th>file_name</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>entity_submitter_id</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>gender</th>\n",
       "      <th>tumor_stage</th>\n",
       "      <th>tumor_grade</th>\n",
       "      <th>days_to_death</th>\n",
       "      <th>days_to_last_follow_up</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>disease_type</th>\n",
       "      <th>project_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-ZH-A8Y1-01A</th>\n",
       "      <td>9e76b112-9c90-4310-9e0b-643519d1b8c8</td>\n",
       "      <td>jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-ZH-A8Y1-01A</td>\n",
       "      <td>TCGA-ZH-A8Y1-01A-11D-A418-05</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>27279</td>\n",
       "      <td>female</td>\n",
       "      <td>stage iva</td>\n",
       "      <td>not reported</td>\n",
       "      <td>385.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA36-01A</th>\n",
       "      <td>8c977e5c-a9a8-4abb-b9f8-61f74dddbb16</td>\n",
       "      <td>jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-W5-AA36-01A</td>\n",
       "      <td>TCGA-W5-AA36-01A-11D-A418-05</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>18882</td>\n",
       "      <td>female</td>\n",
       "      <td>stage iv</td>\n",
       "      <td>not reported</td>\n",
       "      <td>1402.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA2I-11A</th>\n",
       "      <td>77dd33d4-b399-43d0-9ac8-b4798c481294</td>\n",
       "      <td>jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-W5-AA2I-11A</td>\n",
       "      <td>TCGA-W5-AA2I-11A-11D-A418-05</td>\n",
       "      <td>Solid Tissue Normal</td>\n",
       "      <td>24388</td>\n",
       "      <td>male</td>\n",
       "      <td>stage i</td>\n",
       "      <td>not reported</td>\n",
       "      <td>1939.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA2W-01A</th>\n",
       "      <td>05d8aac6-d53a-485b-9d06-674501e96274</td>\n",
       "      <td>jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-W5-AA2W-01A</td>\n",
       "      <td>TCGA-W5-AA2W-01A-11D-A418-05</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>11438</td>\n",
       "      <td>female</td>\n",
       "      <td>stage iva</td>\n",
       "      <td>not reported</td>\n",
       "      <td>924.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-W5-AA2O-01A</th>\n",
       "      <td>bb8db31e-d4a8-48db-9f20-c39a9e24de92</td>\n",
       "      <td>jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>TCGA-W5-AA2O-01A</td>\n",
       "      <td>TCGA-W5-AA2O-01A-11D-A418-05</td>\n",
       "      <td>Primary Tumor</td>\n",
       "      <td>21118</td>\n",
       "      <td>male</td>\n",
       "      <td>stage i</td>\n",
       "      <td>not reported</td>\n",
       "      <td>640.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>dead</td>\n",
       "      <td>Cholangiocarcinoma</td>\n",
       "      <td>TCGA-CHOL</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               file_id  \\\n",
       "TCGA-ZH-A8Y1-01A  9e76b112-9c90-4310-9e0b-643519d1b8c8   \n",
       "TCGA-W5-AA36-01A  8c977e5c-a9a8-4abb-b9f8-61f74dddbb16   \n",
       "TCGA-W5-AA2I-11A  77dd33d4-b399-43d0-9ac8-b4798c481294   \n",
       "TCGA-W5-AA2W-01A  05d8aac6-d53a-485b-9d06-674501e96274   \n",
       "TCGA-W5-AA2O-01A  bb8db31e-d4a8-48db-9f20-c39a9e24de92   \n",
       "\n",
       "                                                          file_name  \\\n",
       "TCGA-ZH-A8Y1-01A  jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-W5-AA36-01A  jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-W5-AA2I-11A  jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-W5-AA2W-01A  jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...   \n",
       "TCGA-W5-AA2O-01A  jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...   \n",
       "\n",
       "                      submitter_id           entity_submitter_id  \\\n",
       "TCGA-ZH-A8Y1-01A  TCGA-ZH-A8Y1-01A  TCGA-ZH-A8Y1-01A-11D-A418-05   \n",
       "TCGA-W5-AA36-01A  TCGA-W5-AA36-01A  TCGA-W5-AA36-01A-11D-A418-05   \n",
       "TCGA-W5-AA2I-11A  TCGA-W5-AA2I-11A  TCGA-W5-AA2I-11A-11D-A418-05   \n",
       "TCGA-W5-AA2W-01A  TCGA-W5-AA2W-01A  TCGA-W5-AA2W-01A-11D-A418-05   \n",
       "TCGA-W5-AA2O-01A  TCGA-W5-AA2O-01A  TCGA-W5-AA2O-01A-11D-A418-05   \n",
       "\n",
       "                          sample_type  age_at_diagnosis  gender tumor_stage  \\\n",
       "TCGA-ZH-A8Y1-01A        Primary Tumor             27279  female   stage iva   \n",
       "TCGA-W5-AA36-01A        Primary Tumor             18882  female    stage iv   \n",
       "TCGA-W5-AA2I-11A  Solid Tissue Normal             24388    male     stage i   \n",
       "TCGA-W5-AA2W-01A        Primary Tumor             11438  female   stage iva   \n",
       "TCGA-W5-AA2O-01A        Primary Tumor             21118    male     stage i   \n",
       "\n",
       "                   tumor_grade  days_to_death  days_to_last_follow_up  \\\n",
       "TCGA-ZH-A8Y1-01A  not reported          385.0                     NaN   \n",
       "TCGA-W5-AA36-01A  not reported         1402.0                     NaN   \n",
       "TCGA-W5-AA2I-11A  not reported         1939.0                     NaN   \n",
       "TCGA-W5-AA2W-01A  not reported          924.0                     NaN   \n",
       "TCGA-W5-AA2O-01A  not reported          640.0                     NaN   \n",
       "\n",
       "                 vital_status        disease_type project_id  \n",
       "TCGA-ZH-A8Y1-01A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA36-01A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA2I-11A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA2W-01A         dead  Cholangiocarcinoma  TCGA-CHOL  \n",
       "TCGA-W5-AA2O-01A         dead  Cholangiocarcinoma  TCGA-CHOL  "
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "###########################################################################\n",
    "###                           Other Functions                           ###\n",
    "###########################################################################\n",
    "\n",
    "### get manifest file\n",
    "\n",
    "def getManifest(project, data_category, data_type, workflow_type):\n",
    "    \n",
    "    files_endpt = \"https://api.gdc.cancer.gov/files\"\n",
    "\n",
    "    # This set of filters is nested under an 'and' operator.\n",
    "    filters = {\n",
    "        \"op\": \"and\",\n",
    "        \"content\":[\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"cases.project.project_id\",\n",
    "                \"value\": [project]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"files.data_category\",\n",
    "                \"value\": [data_category]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"files.data_type\",\n",
    "                \"value\": [data_type]\n",
    "                }\n",
    "            },\n",
    "            {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":{\n",
    "                \"field\": \"files.analysis.workflow_type\",\n",
    "                \"value\": [workflow_type]\n",
    "                }\n",
    "            }\n",
    "        ]\n",
    "    }\n",
    "    \n",
    "    # A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "    params = {\n",
    "        \"filters\": filters,\n",
    "        #\"fields\": fields,\n",
    "        #\"format\": \"TSV\",\n",
    "        \"size\": \"10000\",\n",
    "        \"return_type\": 'manifest'\n",
    "        }\n",
    "    \n",
    "    # The parameters are passed to 'json' rather than 'params' in this case\n",
    "    response = requests.post(files_endpt, headers = {\"Content-Type\": \"application/json\"}, json = params)\n",
    "    \n",
    "    df = pd.read_csv(io.StringIO(response.content.decode(\"utf-8\")), sep='\\t')\n",
    "    \n",
    "    return (df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "manifest = getManifest(project = 'TCGA-CHOL', \n",
    "                    data_category = 'DNA Methylation', \n",
    "                    data_type = 'Methylation Beta Value', \n",
    "                    workflow_type = 'Liftover')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>id</th>\n",
       "      <th>filename</th>\n",
       "      <th>md5</th>\n",
       "      <th>size</th>\n",
       "      <th>state</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9e76b112-9c90-4310-9e0b-643519d1b8c8</td>\n",
       "      <td>jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>1c3969ae227bb2571bd638610fa113ed</td>\n",
       "      <td>141181949</td>\n",
       "      <td>live</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8c977e5c-a9a8-4abb-b9f8-61f74dddbb16</td>\n",
       "      <td>jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>a59a932df97a3972a34a5db8089c6444</td>\n",
       "      <td>141230779</td>\n",
       "      <td>live</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>77dd33d4-b399-43d0-9ac8-b4798c481294</td>\n",
       "      <td>jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>3fbfa356cb84d9aba5fb87d10231fd15</td>\n",
       "      <td>141222861</td>\n",
       "      <td>live</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>05d8aac6-d53a-485b-9d06-674501e96274</td>\n",
       "      <td>jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>22761b47e8dcdc4a1de79c9a75c2900e</td>\n",
       "      <td>141281134</td>\n",
       "      <td>live</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>bb8db31e-d4a8-48db-9f20-c39a9e24de92</td>\n",
       "      <td>jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...</td>\n",
       "      <td>0c5bb44a7c780ace5825ab6da0fe9ef1</td>\n",
       "      <td>141287936</td>\n",
       "      <td>live</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     id  \\\n",
       "0  9e76b112-9c90-4310-9e0b-643519d1b8c8   \n",
       "1  8c977e5c-a9a8-4abb-b9f8-61f74dddbb16   \n",
       "2  77dd33d4-b399-43d0-9ac8-b4798c481294   \n",
       "3  05d8aac6-d53a-485b-9d06-674501e96274   \n",
       "4  bb8db31e-d4a8-48db-9f20-c39a9e24de92   \n",
       "\n",
       "                                            filename  \\\n",
       "0  jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...   \n",
       "1  jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...   \n",
       "2  jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...   \n",
       "3  jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...   \n",
       "4  jhu-usc.edu_CHOL.HumanMethylation450.1.lvl-3.T...   \n",
       "\n",
       "                                md5       size state  \n",
       "0  1c3969ae227bb2571bd638610fa113ed  141181949  live  \n",
       "1  a59a932df97a3972a34a5db8089c6444  141230779  live  \n",
       "2  3fbfa356cb84d9aba5fb87d10231fd15  141222861  live  \n",
       "3  22761b47e8dcdc4a1de79c9a75c2900e  141281134  live  \n",
       "4  0c5bb44a7c780ace5825ab6da0fe9ef1  141287936  live  "
      ]
     },
     "execution_count": 203,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "manifest.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "manifest.to_csv('manifest.txt', header=True, index=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 169,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
