{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of protein coding genes: 19961\n",
      "Number of long non-coding genes: 14672\n",
      "Number of other non-coding genes: 23419\n"
     ]
    }
   ],
   "source": [
    "databases = '/home/ruidong/Documents/Research/Cancer_Genomics/Data/Databases/'\n",
    "from collections import OrderedDict\n",
    "import re\n",
    "from operator import itemgetter\n",
    "\n",
    "#################################\n",
    "## always use new version for id conversion\n",
    "## but also report genes that are not in this version\n",
    "\n",
    "\n",
    "proteinCodingSymbol = OrderedDict()\n",
    "longNonCodingSymbol = OrderedDict()\n",
    "otherNonCodingSymbol = OrderedDict()\n",
    "ensemblToSymbol= OrderedDict()\n",
    "\n",
    "lnc = ['3prime_overlapping_ncrna', 'ambiguous_orf', 'antisense', 'lincRNA', 'ncrna_host', \n",
    "       'non_coding', 'processed_transcript', 'retained_intron', 'sense_intronic', 'sense_overlapping']\n",
    "\n",
    "with open(databases+'Gene_ID_conversion/Homo_sapiens.GRCh38.86.gff3.clean.txt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        geneID = lst[0].split('.')[0]\n",
    "        if lst[2] == 'protein_coding':\n",
    "            proteinCodingSymbol[geneID] = line\n",
    "            ensemblToSymbol[geneID] = line\n",
    "        elif lst[2] in lnc:\n",
    "            longNonCodingSymbol[geneID] = line\n",
    "            ensemblToSymbol[geneID] = line\n",
    "        else:\n",
    "            otherNonCodingSymbol[geneID] = line\n",
    "            ensemblToSymbol[geneID] = line\n",
    "            \n",
    "            \n",
    "print ('Number of protein coding genes: %d' %len(proteinCodingSymbol))\n",
    "print ('Number of long non-coding genes: %d' %len(longNonCodingSymbol))\n",
    "print ('Number of other non-coding genes: %d' %len(otherNonCodingSymbol))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "########################################\n",
    "##### GRCh38.86\n",
    "databases = '/home/ruidong/Documents/Research/Cancer_Genomics/Data/Databases/'\n",
    "from collections import OrderedDict\n",
    "import re\n",
    "from operator import itemgetter\n",
    "\n",
    "fh = open(databases+'Gene_ID_conversion/Homo_sapiens.GRCh37.87.gtf.clean.txt', 'wt')\n",
    "fh.write('\\t'.join(['geneID', 'geneName', 'biotype', 'geneSource' 'description'])+'\\n')\n",
    "\n",
    "with open(databases+'Gene_ID_conversion/Homo_sapiens.GRCh37.87.gtf') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        if line.startswith('#'):\n",
    "            continue\n",
    "        lst = line.split('\\t')\n",
    "        if lst[2] == 'gene':\n",
    "            geneInfo = lst[-1].split(';')\n",
    "            \n",
    "            geneId = geneInfo[0].split('\"')[1]\n",
    "            geneVersion = geneInfo[1].split('\"')[1]\n",
    "            geneEnsembl = geneId+'.'+geneVersion\n",
    "            geneName = geneInfo[2].split('\"')[1]\n",
    "            geneSource = geneInfo[3].split('\"')[1]\n",
    "            geneType = geneInfo[4].split('\"')[1]\n",
    "            \n",
    "            if geneName in description.keys():\n",
    "                geneDescription = description[geneName]\n",
    "            else:\n",
    "                geneDescription = 'NA'\n",
    "            \n",
    "            geneSummary = '\\t'.join([geneEnsembl, geneName, geneType, geneSource, geneDescription])\n",
    "            fh.write(geneSummary+'\\n')\n",
    "            \n",
    "            if geneType == 'protein_coding':\n",
    "                proteinCodingSymbol[geneEnsembl] = geneSummary\n",
    "            else:\n",
    "                nonCodingSymbol[geneEnsembl] = geneSummary\n",
    "            \n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "########################################\n",
    "##### GRCh38.86\n",
    "\n",
    "import re\n",
    "fh = open(databases+'Gene_ID_conversion/Homo_sapiens.GRCh38.86.gff3.clean.txt', 'wt')\n",
    "fh.write('geneID\\tgeneName\\tbiotype\\tdescription\\n')\n",
    "with open(databases+'Gene_ID_conversion/Homo_sapiens.GRCh38.86.gff3') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        if line.startswith('#'):\n",
    "            continue\n",
    "        lst = line.split('\\t')\n",
    "        if lst[-1].startswith('ID=gene:ENSG'):\n",
    "            geneInfo = lst[-1].split(';')\n",
    "                    \n",
    "            if re.search('description', lst[-1]):\n",
    "                geneId = geneInfo[4].split('=')[-1]+'.'+geneInfo[-1].split('=')[-1]\n",
    "                geneName = geneInfo[1].split('=')[-1]\n",
    "                biotype = geneInfo[2].split('=')[-1]\n",
    "                description = geneInfo[3].split('=')[-1]\n",
    "            else:\n",
    "                geneId = geneInfo[3].split('=')[-1]+'.'+geneInfo[-1].split('=')[-1]\n",
    "                geneName = geneInfo[1].split('=')[-1]\n",
    "                biotype = geneInfo[2].split('=')[-1]\n",
    "                description = 'NA'\n",
    "            fh.write('\\t'.join([geneId, geneName, biotype, description])+'\\n')\n",
    "            \n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "########################################\n",
    "##### GRCh38.79\n",
    "\n",
    "from collections import OrderedDict\n",
    "\n",
    "description = OrderedDict()\n",
    "\n",
    "with open(databases+'Gene_ID_conversion/Homo_sapiens.GRCh38.86.gff3.clean.txt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        if line.startswith('geneID'):\n",
    "            continue\n",
    "            \n",
    "        if lst[-1] != 'NA':\n",
    "            description[lst[1]] = lst[-1].split(' [')[0]\n",
    "        else:\n",
    "            description[lst[1]] = 'NA'\n",
    "\n",
    "proteinCodingSymbol = OrderedDict()\n",
    "nonCodingSymbol = OrderedDict()\n",
    "\n",
    "fh = open(databases+'Gene_ID_conversion/Homo_sapiens.GRCh38.79.clean.gtf', 'wt')\n",
    "fh.write('\\t'.join(['geneID', 'geneName', 'biotype', 'geneSource' 'description'])+'\\n')\n",
    "\n",
    "with open(databases+'Gene_ID_conversion/Homo_sapiens.GRCh38.79.gtf') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        if line.startswith('#'):\n",
    "            continue\n",
    "        lst = line.split('\\t')\n",
    "        if lst[2] == 'gene':\n",
    "            geneInfo = lst[-1].split(';')\n",
    "            \n",
    "            geneId = geneInfo[0].split('\"')[1]\n",
    "            geneVersion = geneInfo[1].split('\"')[1]\n",
    "            geneEnsembl = geneId+'.'+geneVersion\n",
    "            geneName = geneInfo[2].split('\"')[1]\n",
    "            geneSource = geneInfo[3].split('\"')[1]\n",
    "            geneType = geneInfo[4].split('\"')[1]\n",
    "            \n",
    "            if geneName in description.keys():\n",
    "                geneDescription = description[geneName]\n",
    "            else:\n",
    "                geneDescription = 'NA'\n",
    "            \n",
    "            geneSummary = '\\t'.join([geneEnsembl, geneName, geneType, geneSource, geneDescription])\n",
    "            fh.write(geneSummary+'\\n')\n",
    "            \n",
    "            if geneType == 'protein_coding':\n",
    "                proteinCodingSymbol[geneEnsembl] = geneSummary\n",
    "            else:\n",
    "                nonCodingSymbol[geneEnsembl] = geneSummary\n",
    "            \n",
    "fh.close()\n",
    "\n",
    "print ('Number of protein coding genes: %d' %len(proteinCodingSymbol))\n",
    "print ('Number of long non-coding genes: %d' %len(nonCodingSymbol))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "fh1 = open(databases+'Gene_ID_conversion/retired_ensembl_id.txt', 'wt')\n",
    "fh2 = open(databases+'Gene_ID_conversion/retired_gene_name.txt', 'wt')\n",
    "version79 = OrderedDict()\n",
    "version86 = OrderedDict()\n",
    "with open(databases+'Gene_ID_conversion/Homo_sapiens.GRCh38.79.clean.gtf') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        gene = lst[0].split('.')[0]\n",
    "        \n",
    "        version79[gene] = lst[1]\n",
    "        \n",
    "        \n",
    "with open(databases+'Gene_ID_conversion/Homo_sapiens.GRCh38.86.gff3.clean.txt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        gene = lst[0].split('.')[0]\n",
    "        \n",
    "        version86[gene] = lst[1]\n",
    "        \n",
    "for ke, val in version79.items():\n",
    "    if ke not in version86.keys():\n",
    "        fh1.write(ke+'\\t'+val+'\\n')\n",
    "    else:\n",
    "        if val != version86[ke]:\n",
    "            fh2.write('\\t'.join([ke,val,version86[ke]])+'\\n')\n",
    "fh1.close()\n",
    "fh2.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#################################################\n",
    "#### Organize Human miRNAs data from miRBase ####\n",
    "#################################################\n",
    "os.chdir('/home/ruidong/Documents/Research/Cancer_Genomics/Data/Import Database/')\n",
    "\n",
    "fh = open('miRBase.hsa.v21.clean.txt', 'wt')\n",
    "stemLoop21 = OrderedDict()\n",
    "\n",
    "with open('hsa.v21.gff3.txt', 'rt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        if line.startswith('#'):\n",
    "            continue\n",
    "        lst = line.split('\\t')\n",
    "        stemLoopInfo = lst[8].split(';')\n",
    "        if lst[2] == 'miRNA_primary_transcript':\n",
    "            stemLoopName = stemLoopInfo[2][5:]+'\\t'+stemLoopInfo[0][3:]\n",
    "            stemLoop21[stemLoopName] = []\n",
    "            continue\n",
    "        maturemiRNA = stemLoopInfo[2][5:]+'\\t'+stemLoopInfo[0][3:]\n",
    "        stemLoop21[stemLoopName].append(maturemiRNA)\n",
    "\n",
    "sortedstemLoop = sorted(stemLoop21.items(), key = itemgetter(0))\n",
    "\n",
    "for ke, val in sortedstemLoop:\n",
    "    for subval in val:\n",
    "        fh.write('\\t'.join([ke,subval])+'\\n')\n",
    "        \n",
    "fh.close()\n",
    "\n",
    "\n",
    "fh = open('miRBase.hsa.v20.clean.txt', 'wt')\n",
    "stemLoop20 = OrderedDict()\n",
    "\n",
    "with open('hsa.v20.gff3.txt', 'rt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        if line.startswith('#'):\n",
    "            continue\n",
    "        lst = line.split('\\t')\n",
    "        stemLoopInfo = lst[8].split(';')\n",
    "        if lst[2] == 'miRNA_primary_transcript':\n",
    "            stemLoopName = stemLoopInfo[2][5:]+'\\t'+stemLoopInfo[0][3:]\n",
    "            stemLoop20[stemLoopName] = []\n",
    "            continue\n",
    "        maturemiRNA = stemLoopInfo[2][5:]+'\\t'+stemLoopInfo[0][3:]\n",
    "        stemLoop20[stemLoopName].append(maturemiRNA)\n",
    "\n",
    "sortedstemLoop = sorted(stemLoop20.items(), key = itemgetter(0))\n",
    "\n",
    "for ke, val in sortedstemLoop:\n",
    "    for subval in val:\n",
    "        fh.write('\\t'.join([ke,subval])+'\\n')\n",
    "        \n",
    "fh.close()\n",
    "\n",
    "fh = open('miRBase.hsa.v19.clean.txt', 'wt')\n",
    "stemLoop19 = OrderedDict()\n",
    "\n",
    "with open('hsa.v19.gff3.txt', 'rt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        if line.startswith('#'):\n",
    "            continue\n",
    "        lst = line.split('\\t')\n",
    "        stemLoopInfo = lst[8].split(';')\n",
    "        if lst[2] == 'miRNA_primary_transcript':\n",
    "            stemLoopName = stemLoopInfo[2][5:]+'\\t'+stemLoopInfo[0][3:]\n",
    "            stemLoop19[stemLoopName] = []\n",
    "            continue\n",
    "        maturemiRNA = stemLoopInfo[2][5:]+'\\t'+stemLoopInfo[0][3:]\n",
    "        stemLoop19[stemLoopName].append(maturemiRNA)\n",
    "\n",
    "sortedstemLoop = sorted(stemLoop19.items(), key = itemgetter(0))\n",
    "\n",
    "for ke, val in sortedstemLoop:\n",
    "    for subval in val:\n",
    "        fh.write('\\t'.join([ke,subval])+'\\n')\n",
    "        \n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "################################################\n",
    "#### Organize Human miRNAs families miRBase ####\n",
    "################################################\n",
    "fh = open(databases+'miRBase/miFam.clean.dat','wt')\n",
    "fh.write('\\t'.join(['AC', 'ID', 'MI'])+'\\n')\n",
    "\n",
    "with open(databases+'miRBase/miFam.dat', 'rt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split(' ')\n",
    "        \n",
    "        if line.startswith('AC'):\n",
    "            ac = lst[-1]\n",
    "        elif line.startswith('ID'):\n",
    "            family = lst[-1]\n",
    "        \n",
    "        if lst[-1].startswith('hsa'):\n",
    "            fh.write('\\t'.join([ac, family, lst[-1]])+'\\n')\n",
    "            \n",
    "fh.close()\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "########################\n",
    "#### TargetScan7.1  ####\n",
    "########################\n",
    "fh1 = open(databases+'miRNA_target_prediction/summary_Predicted_Targets_Context_Scores.default_predictions_95.txt', 'wt')\n",
    "fh2 = open(databases+'miRNA_target_prediction/summary2_Predicted_Targets_Context_Scores.default_predictions_95.txt', 'wt')\n",
    "\n",
    "targetscan = OrderedDict()\n",
    "\n",
    "with open(databases+'miRNA_target_prediction/Predicted_Targets_Context_Scores.default_predictions.txt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        if line.startswith('Gene'):\n",
    "            continue\n",
    "        elif not lst[4].startswith('hsa'):\n",
    "            continue\n",
    "        elif int(lst[-1]) < 95:\n",
    "            continue\n",
    "            \n",
    "        mirnas = '-'.join(lst[4].split('-')[0:3]) + '\\t' + lst[4]\n",
    "        \n",
    "        if mirnas not in targetscan.keys():\n",
    "            targetscan[mirnas] = []\n",
    "        gene = lst[0].split('.')[0]\n",
    "        \n",
    "        if gene in ensemblToSymbol.keys():\n",
    "            targetscan[mirnas].append(ensemblToSymbol[gene].split('\\t')[1])\n",
    "        else:\n",
    "            targetscan[mirnas].append('*'+lst[1])\n",
    "            \n",
    "sortedTargetScan = sorted(targetscan.items(), key=itemgetter(0))\n",
    "\n",
    "for ke, val in sortedTargetScan:\n",
    "    fh1.write(ke+'\\t'+','.join(val)+'\\n')\n",
    "    for subv in val:\n",
    "        fh2.write('\\t'.join([ke, subv])+'\\n')\n",
    "        \n",
    "fh1.close()\n",
    "fh2.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ENSG00000167173.14\tC15orf39\tENST00000567617.1\t9544\tmml-miR-149-5p\t3\t71\t78\t-0.214\t92\t-0.214\t94\n",
      "\n"
     ]
    }
   ],
   "source": [
    "with open(databases+'miRNA_target_prediction/Predicted_Targets_Context_Scores.default_predictions.txt') as f:\n",
    "    for line in f:\n",
    "        if re.search('C15orf39', line):\n",
    "            print (line)\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "########################################\n",
    "######## GO\n",
    "\n",
    "goMF = OrderedDict()\n",
    "goBP = OrderedDict()\n",
    "goCC = OrderedDict()\n",
    "\n",
    "\n",
    "with open(databases+'ensembl_GO.txt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        \n",
    "        if line.startswith('Gene'):\n",
    "            continue\n",
    "            \n",
    "        elif len(lst) == 1:\n",
    "            continue\n",
    "            \n",
    "        goterm = lst[2]+'\\t'+lst[3]\n",
    "            \n",
    "        if lst[1] == 'molecular_function':\n",
    "            if goterm not in goMF.keys():\n",
    "                goMF[goterm] = []\n",
    "            \n",
    "            if lst[0] not in goMF[goterm]:\n",
    "                goMF[goterm].append(lst[0])\n",
    "                \n",
    "        elif lst[1] == 'biological_process':\n",
    "            if goterm not in goBP.keys():\n",
    "                goBP[goterm] = []\n",
    "            \n",
    "            if lst[0] not in goBP[goterm]:\n",
    "                goBP[goterm].append(lst[0])\n",
    "                \n",
    "        \n",
    "        \n",
    "        elif lst[1] == 'cellular_component':\n",
    "            if goterm not in goCC.keys():\n",
    "                goCC[goterm] = []\n",
    "            \n",
    "            if lst[0] not in goCC[goterm]:\n",
    "                goCC[goterm].append(lst[0])\n",
    "                \n",
    "fh = open(databases+'ensembl_GO.cleaned.txt', 'wt')\n",
    "\n",
    "for ke,val in goMF.items():\n",
    "    fh.write('molecular_function'+'\\t'+ke+'\\t'+';'.join(val)+'\\n')\n",
    "    \n",
    "for ke,val in goBP.items():\n",
    "    fh.write('biological_process'+'\\t'+ke+'\\t'+';'.join(val)+'\\n')\n",
    "    \n",
    "for ke,val in goCC.items():\n",
    "    fh.write('cellular_component'+'\\t'+ke+'\\t'+';'.join(val)+'\\n')\n",
    "    \n",
    "    \n",
    "fh.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19989\n"
     ]
    }
   ],
   "source": [
    "########################################\n",
    "######## GO\n",
    "\n",
    "\n",
    "go = OrderedDict()\n",
    "with open(databases+'ensembl_GO.txt') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "        lst = line.split('\\t')\n",
    "        \n",
    "        if line.startswith('Gene'):\n",
    "            continue\n",
    "            \n",
    "        elif len(lst) == 1:\n",
    "            continue\n",
    "            \n",
    "        elif lst[1] == 'cellular_component':\n",
    "            if lst[0] not in go.keys():\n",
    "                go[lst[0]] = 1\n",
    "                \n",
    "                \n",
    "                \n",
    "print (len(go))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "########################################\n",
    "######## GO\n",
    "\n",
    "\n",
    "fh = open(databases+'GO/go-basic-02142017.obo.txt', 'wt')\n",
    "fh.write('\\t'.join(['Domain', 'goID', 'goName', 'Definition', 'is_a'])+'\\n')\n",
    "\n",
    "\n",
    "go = OrderedDict()\n",
    "go['biological_process'] = []\n",
    "go['molecular_function'] = []\n",
    "go['cellular_component'] = []\n",
    "\n",
    "switch = 0\n",
    "\n",
    "with open(databases+'GO/go-basic.obo') as f:\n",
    "    for line in f:\n",
    "        line = line.rstrip()\n",
    "\n",
    "        if line.startswith('[Term]'):\n",
    "            if switch == 0:\n",
    "                isA = []\n",
    "                switch = 1\n",
    "                continue\n",
    "            else:\n",
    "                go[goNamespace].append('\\t'.join([goID, goName, goDef, ';'.join(isA)]))\n",
    "                isA = []\n",
    "            \n",
    "        elif line.startswith('id:'):\n",
    "            goID = re.sub('id:\\s+', '', line)\n",
    "\n",
    "        elif line.startswith('name:'):\n",
    "            goName = re.sub('name:\\s+', '', line)\n",
    "            \n",
    "        elif line.startswith('namespace:'):\n",
    "            goNamespace = re.sub('namespace:\\s+', '', line)\n",
    "            \n",
    "        elif line.startswith('def:'):\n",
    "            goDef = re.sub('def:\\s+', '', line)\n",
    "            goDef = re.sub('\"', '', goDef)\n",
    "            \n",
    "        elif line.startswith('is_a'):\n",
    "            isA.append(re.search('GO:\\d+',line).group(0))\n",
    "            \n",
    "        elif line.startswith('[Typedef]'):\n",
    "            go[goNamespace].append('\\t'.join([goID, goName, goDef, ';'.join(isA)]))\n",
    "            break\n",
    "            \n",
    "            \n",
    "            \n",
    "for ke, val in go.items():\n",
    "    for subv in val:\n",
    "        fh.write(ke+'\\t'+subv+'\\n')\n",
    "        \n",
    "\n",
    "fh.close()\n",
    "        \n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [Root]",
   "language": "python",
   "name": "Python [Root]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
